--- a/net/minecraft/entity/projectile/ExplosiveProjectileEntity.java
+++ b/net/minecraft/entity/projectile/ExplosiveProjectileEntity.java
@@ -70,7 +70,7 @@
          }
 
          HitResult raytraceresult = ProjectileUtil.getCollision(this, this::method_26823);
-         if (raytraceresult.getType() != HitResult.Type.MISS) {
+         if (raytraceresult.getType() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.onCollision(raytraceresult);
          }
 
