--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -157,6 +157,13 @@
 
    }
 
+   @Override
+   protected void onCollision(HitResult result) {
+      if (result.getType() == HitResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.onCollision(result);
+      }
+   }
+
    private void explodeAndRemove() {
       this.world.sendEntityStatus(this, (byte)17);
       this.explode();
