--- a/net/minecraft/entity/projectile/FireballEntity.java
+++ b/net/minecraft/entity/projectile/FireballEntity.java
@@ -34,7 +34,7 @@
    protected void onCollision(HitResult result) {
       super.onCollision(result);
       if (!this.world.isClient) {
-         boolean flag = this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING);
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.getOwner());
          this.world.createExplosion((Entity)null, this.getX(), this.getY(), this.getZ(), (float)this.explosionPower, flag, flag ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE);
          this.remove();
       }
