--- a/net/minecraft/entity/projectile/FishingBobberEntity.java
+++ b/net/minecraft/entity/projectile/FishingBobberEntity.java
@@ -235,8 +235,8 @@
    private boolean removeIfInvalid(PlayerEntity p_234600_1_) {
       ItemStack itemstack = p_234600_1_.getMainHandStack();
       ItemStack itemstack1 = p_234600_1_.getOffHandStack();
-      boolean flag = itemstack.getItem() == Items.FISHING_ROD;
-      boolean flag1 = itemstack1.getItem() == Items.FISHING_ROD;
+      boolean flag = itemstack.getItem() instanceof net.minecraft.item.FishingRodItem;
+      boolean flag1 = itemstack1.getItem() instanceof net.minecraft.item.FishingRodItem;
       if (!p_234600_1_.removed && p_234600_1_.isAlive() && (flag || flag1) && !(this.squaredDistanceTo(p_234600_1_) > 1024.0D)) {
          return false;
       } else {
@@ -302,7 +302,7 @@
             double d1 = (double)((float)MathHelper.floor(this.getY()) + 1.0F);
             double d2 = this.getZ() + (double)(f2 * (float)this.fishTravelCountdown * 0.1F);
             BlockState blockstate = serverworld.getBlockState(new BlockPos(d0, d1 - 1.0D, d2));
-            if (blockstate.isOf(Blocks.WATER)) {
+            if (serverworld.getBlockState(new BlockPos((int)d0, (int)d1 - 1, (int)d2)).getMaterial() == net.minecraft.block.Material.WATER) {
                if (this.random.nextFloat() < 0.15F) {
                   serverworld.spawnParticles(ParticleTypes.BUBBLE, d0, d1 - (double)0.1F, d2, 1, (double)f1, 0.1D, (double)f2, 0.0D);
                }
@@ -338,7 +338,7 @@
             double d5 = (double)((float)MathHelper.floor(this.getY()) + 1.0F);
             double d6 = this.getZ() + (double)(MathHelper.cos(f6) * f7 * 0.1F);
             BlockState blockstate1 = serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6));
-            if (blockstate1.isOf(Blocks.WATER)) {
+            if (serverworld.getBlockState(new BlockPos(d4, d5 - 1.0D, d6)).getMaterial() == net.minecraft.block.Material.WATER) {
                serverworld.spawnParticles(ParticleTypes.SPLASH, d4, d5, d6, 2 + this.random.nextInt(2), (double)0.1F, 0.0D, (double)0.1F, 0.0D);
             }
          }
@@ -409,6 +409,7 @@
       PlayerEntity playerentity = this.getPlayerOwner();
       if (!this.world.isClient && playerentity != null) {
          int i = 0;
+         net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
          if (this.hookedEntity != null) {
             this.pullHookedEntity();
             Criteria.FISHING_ROD_HOOKED.trigger((ServerPlayerEntity)playerentity, p_146034_1_, this, Collections.emptyList());
@@ -416,8 +417,15 @@
             i = this.hookedEntity instanceof ItemEntity ? 3 : 5;
          } else if (this.hookCountdown > 0) {
             LootContext.Builder lootcontext$builder = (new LootContext.Builder((ServerWorld)this.world)).parameter(LootContextParameters.ORIGIN, this.getPos()).parameter(LootContextParameters.TOOL, p_146034_1_).parameter(LootContextParameters.THIS_ENTITY, this).random(this.random).luck((float)this.luckOfTheSeaLevel + playerentity.getLuck());
+            lootcontext$builder.parameter(LootContextParameters.KILLER_ENTITY, this.getOwner()).parameter(LootContextParameters.THIS_ENTITY, this);
             LootTable loottable = this.world.getServer().getLootManager().getTable(LootTables.FISHING_GAMEPLAY);
             List<ItemStack> list = loottable.generateLoot(lootcontext$builder.build(LootContextTypes.FISHING));
+            event = new net.minecraftforge.event.entity.player.ItemFishedEvent(list, this.onGround ? 2 : 1, this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            if (event.isCanceled()) {
+               this.remove();
+               return event.getRodDamage();
+            }
             Criteria.FISHING_ROD_HOOKED.trigger((ServerPlayerEntity)playerentity, p_146034_1_, this, list);
 
             for(ItemStack itemstack : list) {
@@ -442,7 +450,7 @@
          }
 
          this.remove();
-         return i;
+         return event == null ? i : event.getRodDamage();
       } else {
          return 0;
       }
@@ -469,8 +477,9 @@
       return false;
    }
 
-   public void remove() {
-      super.remove();
+   @Override
+   public void remove(boolean keepData) {
+      super.remove(keepData);
       PlayerEntity playerentity = this.getPlayerOwner();
       if (playerentity != null) {
          playerentity.fishHook = null;
