--- a/net/minecraft/entity/projectile/WitherSkullEntity.java
+++ b/net/minecraft/entity/projectile/WitherSkullEntity.java
@@ -50,7 +50,7 @@
    }
 
    public float getEffectiveExplosionResistance(Explosion explosionIn, BlockView worldIn, BlockPos pos, BlockState blockStateIn, FluidState fluidState, float explosionPower) {
-      return this.isCharged() && WitherEntity.canDestroy(blockStateIn) ? Math.min(0.8F, explosionPower) : explosionPower;
+      return this.isCharged() && blockStateIn.canEntityDestroy(worldIn, pos, this) ? Math.min(0.8F, explosionPower) : explosionPower;
    }
 
    protected void onEntityHit(EntityHitResult p_213868_1_) {
@@ -92,7 +92,7 @@
    protected void onCollision(HitResult result) {
       super.onCollision(result);
       if (!this.world.isClient) {
-         Explosion.DestructionType explosion$mode = this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
+         Explosion.DestructionType explosion$mode = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.getOwner()) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
          this.world.createExplosion(this, this.getX(), this.getY(), this.getZ(), 1.0F, false, explosion$mode);
          this.remove();
       }
