--- a/net/minecraft/potion/Potion.java
+++ b/net/minecraft/potion/Potion.java
@@ -7,7 +7,8 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
 
-public class Potion {
+public class Potion extends net.minecraftforge.registries.ForgeRegistryEntry<Potion> {
+   private final net.minecraftforge.common.util.ReverseTagWrapper<Potion> reverseTags = new net.minecraftforge.common.util.ReverseTagWrapper<>(this, () -> net.minecraft.tag.ServerTagManagerHolder.getTagManager().getCustomTypeCollection(net.minecraftforge.registries.ForgeRegistries.POTION_TYPES));
    private final String baseName;
    private final ImmutableList<StatusEffectInstance> effects;
 
@@ -24,6 +25,14 @@
       this.effects = ImmutableList.copyOf(effectsIn);
    }
 
+   public java.util.Set<net.minecraft.util.Identifier> getTags() {
+      return reverseTags.getTagNames();
+   }
+
+   public boolean isIn(net.minecraft.tag.Tag<Potion> tag) {
+      return tag.contains(this);
+   }
+
    public String finishTranslationKey(String prefix) {
       return prefix + (this.baseName == null ? Registry.POTION.getId(this).getPath() : this.baseName);
    }
