--- a/net/minecraft/item/ArmorItem.java
+++ b/net/minecraft/item/ArmorItem.java
@@ -68,7 +68,7 @@
       UUID uuid = MODIFIERS[slot.getEntitySlotId()];
       builder.put(EntityAttributes.GENERIC_ARMOR, new EntityAttributeModifier(uuid, "Armor modifier", (double)this.protection, EntityAttributeModifier.Operation.ADDITION));
       builder.put(EntityAttributes.GENERIC_ARMOR_TOUGHNESS, new EntityAttributeModifier(uuid, "Armor toughness", (double)this.toughness, EntityAttributeModifier.Operation.ADDITION));
-      if (materialIn == ArmorMaterials.NETHERITE) {
+      if (this.knockbackResistance > 0) {
          builder.put(EntityAttributes.GENERIC_KNOCKBACK_RESISTANCE, new EntityAttributeModifier(uuid, "Armor knockback resistance", (double)this.knockbackResistance, EntityAttributeModifier.Operation.ADDITION));
       }
 
