--- a/net/minecraft/item/AxeItem.java
+++ b/net/minecraft/item/AxeItem.java
@@ -22,7 +22,7 @@
    protected static final Map<Block, Block> STRIPPED_BLOCKS = (new Builder<Block, Block>()).put(Blocks.OAK_WOOD, Blocks.STRIPPED_OAK_WOOD).put(Blocks.OAK_LOG, Blocks.STRIPPED_OAK_LOG).put(Blocks.DARK_OAK_WOOD, Blocks.STRIPPED_DARK_OAK_WOOD).put(Blocks.DARK_OAK_LOG, Blocks.STRIPPED_DARK_OAK_LOG).put(Blocks.ACACIA_WOOD, Blocks.STRIPPED_ACACIA_WOOD).put(Blocks.ACACIA_LOG, Blocks.STRIPPED_ACACIA_LOG).put(Blocks.BIRCH_WOOD, Blocks.STRIPPED_BIRCH_WOOD).put(Blocks.BIRCH_LOG, Blocks.STRIPPED_BIRCH_LOG).put(Blocks.JUNGLE_WOOD, Blocks.STRIPPED_JUNGLE_WOOD).put(Blocks.JUNGLE_LOG, Blocks.STRIPPED_JUNGLE_LOG).put(Blocks.SPRUCE_WOOD, Blocks.STRIPPED_SPRUCE_WOOD).put(Blocks.SPRUCE_LOG, Blocks.STRIPPED_SPRUCE_LOG).put(Blocks.WARPED_STEM, Blocks.STRIPPED_WARPED_STEM).put(Blocks.WARPED_HYPHAE, Blocks.STRIPPED_WARPED_HYPHAE).put(Blocks.CRIMSON_STEM, Blocks.STRIPPED_CRIMSON_STEM).put(Blocks.CRIMSON_HYPHAE, Blocks.STRIPPED_CRIMSON_HYPHAE).build();
 
    public AxeItem(ToolMaterial tier, float attackDamageIn, float attackSpeedIn, Item.Settings builder) {
-      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder);
+      super(attackDamageIn, attackSpeedIn, tier, EFFECTIVE_BLOCKS, builder.addToolType(net.minecraftforge.common.ToolType.AXE, tier.getMiningLevel()));
    }
 
    public float getMiningSpeedMultiplier(ItemStack stack, BlockState state) {
@@ -34,12 +34,12 @@
       World world = context.getWorld();
       BlockPos blockpos = context.getBlockPos();
       BlockState blockstate = world.getBlockState(blockpos);
-      Block block = STRIPPED_BLOCKS.get(blockstate.getBlock());
+      BlockState block = blockstate.getToolModifiedState(world, blockpos, context.getPlayer(), context.getStack(), net.minecraftforge.common.ToolType.AXE);
       if (block != null) {
          PlayerEntity playerentity = context.getPlayer();
          world.playSound(playerentity, blockpos, SoundEvents.ITEM_AXE_STRIP, SoundCategory.BLOCKS, 1.0F, 1.0F);
          if (!world.isClient) {
-            world.setBlockState(blockpos, block.getDefaultState().with(PillarBlock.AXIS, blockstate.get(PillarBlock.AXIS)), 11);
+            world.setBlockState(blockpos, block, 11);
             if (playerentity != null) {
                context.getStack().damage(1, playerentity, (p_220040_1_) -> {
                   p_220040_1_.sendToolBreakStatus(context.getHand());
@@ -52,4 +52,10 @@
          return ActionResult.PASS;
       }
    }
+
+   @javax.annotation.Nullable
+   public static BlockState getAxeStrippingState(BlockState originalState) {
+      Block block = STRIPPED_BLOCKS.get(originalState.getBlock());
+      return block != null ? block.getDefaultState().with(PillarBlock.AXIS, originalState.get(PillarBlock.AXIS)) : null;
+   }
 }
