--- a/net/minecraft/item/BoneMealItem.java
+++ b/net/minecraft/item/BoneMealItem.java
@@ -32,7 +32,7 @@
       World world = context.getWorld();
       BlockPos blockpos = context.getBlockPos();
       BlockPos blockpos1 = blockpos.offset(context.getSide());
-      if (useOnFertilizable(context.getStack(), world, blockpos)) {
+      if (applyBonemeal(context.getStack(), world, blockpos, context.getPlayer())) {
          if (!world.isClient) {
             world.syncWorldEvent(2005, blockpos, 0);
          }
@@ -53,8 +53,17 @@
       }
    }
 
+   @Deprecated //Forge: Use Player/Hand version
    public static boolean useOnFertilizable(ItemStack stack, World worldIn, BlockPos pos) {
+      if (worldIn instanceof net.minecraft.server.world.ServerWorld)
+         return applyBonemeal(stack, worldIn, pos, net.minecraftforge.common.util.FakePlayerFactory.getMinecraft((net.minecraft.server.world.ServerWorld)worldIn));
+      return false;
+   }
+
+   public static boolean applyBonemeal(ItemStack stack, World worldIn, BlockPos pos, net.minecraft.entity.player.PlayerEntity player) {
       BlockState blockstate = worldIn.getBlockState(pos);
+      int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, worldIn, pos, blockstate, stack);
+      if (hook != 0) return hook > 0;
       if (blockstate.getBlock() instanceof Fertilizable) {
          Fertilizable igrowable = (Fertilizable)blockstate.getBlock();
          if (igrowable.isFertilizable(worldIn, pos, blockstate, worldIn.isClient)) {
@@ -130,7 +139,7 @@
       }
 
       BlockState blockstate = worldIn.getBlockState(posIn);
-      if (!blockstate.isAir()) {
+      if (!blockstate.isAir(worldIn, posIn)) {
          double d0 = 0.5D;
          double d1;
          if (blockstate.isOf(Blocks.WATER)) {
