--- a/net/minecraft/item/BowItem.java
+++ b/net/minecraft/item/BowItem.java
@@ -24,18 +24,23 @@
          PlayerEntity playerentity = (PlayerEntity)entityLiving;
          boolean flag = playerentity.abilities.creativeMode || EnchantmentHelper.getLevel(Enchantments.INFINITY, stack) > 0;
          ItemStack itemstack = playerentity.getArrowType(stack);
+
+         int i = this.getMaxUseTime(stack) - timeLeft;
+         i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(stack, worldIn, playerentity, i, !itemstack.isEmpty() || flag);
+         if (i < 0) return;
+
          if (!itemstack.isEmpty() || flag) {
             if (itemstack.isEmpty()) {
                itemstack = new ItemStack(Items.ARROW);
             }
 
-            int i = this.getMaxUseTime(stack) - timeLeft;
             float f = getPullProgress(i);
             if (!((double)f < 0.1D)) {
-               boolean flag1 = flag && itemstack.getItem() == Items.ARROW;
+               boolean flag1 = playerentity.abilities.creativeMode || (itemstack.getItem() instanceof ArrowItem && ((ArrowItem)itemstack.getItem()).isInfinite(itemstack, stack, playerentity));
                if (!worldIn.isClient) {
                   ArrowItem arrowitem = (ArrowItem)(itemstack.getItem() instanceof ArrowItem ? itemstack.getItem() : Items.ARROW);
                   PersistentProjectileEntity abstractarrowentity = arrowitem.createArrow(worldIn, itemstack, playerentity);
+                  abstractarrowentity = customArrow(abstractarrowentity);
                   abstractarrowentity.setProperties(playerentity, playerentity.pitch, playerentity.yaw, 0.0F, f * 3.0F, 1.0F);
                   if (f == 1.0F) {
                      abstractarrowentity.setCritical(true);
@@ -100,6 +105,10 @@
    public TypedActionResult<ItemStack> use(World worldIn, PlayerEntity playerIn, Hand handIn) {
       ItemStack itemstack = playerIn.getStackInHand(handIn);
       boolean flag = !playerIn.getArrowType(itemstack).isEmpty();
+
+      TypedActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onArrowNock(itemstack, worldIn, playerIn, handIn, flag);
+      if (ret != null) return ret;
+
       if (!playerIn.abilities.creativeMode && !flag) {
          return TypedActionResult.fail(itemstack);
       } else {
@@ -112,6 +121,10 @@
       return BOW_PROJECTILES;
    }
 
+   public PersistentProjectileEntity customArrow(PersistentProjectileEntity arrow) {
+      return arrow;
+   }
+
    public int getRange() {
       return 15;
    }
