--- a/net/minecraft/item/BucketItem.java
+++ b/net/minecraft/item/BucketItem.java
@@ -31,14 +31,28 @@
 public class BucketItem extends Item {
    private final Fluid fluid;
 
+   // Forge: Use the other constructor that takes a Supplier
+   @Deprecated
    public BucketItem(Fluid containedFluidIn, Item.Settings builder) {
       super(builder);
       this.fluid = containedFluidIn;
+      this.fluidSupplier = containedFluidIn.delegate;
    }
 
+   /**
+    * @param supplier A fluid supplier such as {@link net.minecraftforge.fml.RegistryObject<Fluid>}
+    */
+   public BucketItem(java.util.function.Supplier<? extends Fluid> supplier, Item.Settings builder) {
+      super(builder);
+      this.fluid = null;
+      this.fluidSupplier = supplier;
+   }
+
    public TypedActionResult<ItemStack> use(World worldIn, PlayerEntity playerIn, Hand handIn) {
       ItemStack itemstack = playerIn.getStackInHand(handIn);
       HitResult raytraceresult = raycast(worldIn, playerIn, this.fluid == Fluids.EMPTY ? RaycastContext.FluidHandling.SOURCE_ONLY : RaycastContext.FluidHandling.NONE);
+      TypedActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(playerIn, worldIn, itemstack, raytraceresult);
+      if (ret != null) return ret;
       if (raytraceresult.getType() == HitResult.Type.MISS) {
          return TypedActionResult.pass(itemstack);
       } else if (raytraceresult.getType() != HitResult.Type.BLOCK) {
@@ -55,7 +69,10 @@
                   Fluid fluid = ((FluidDrainable)blockstate1.getBlock()).tryDrainFluid(worldIn, blockpos, blockstate1);
                   if (fluid != Fluids.EMPTY) {
                      playerIn.incrementStat(Stats.USED.getOrCreateStat(this));
-                     playerIn.playSound(fluid.isIn(FluidTags.LAVA) ? SoundEvents.ITEM_BUCKET_FILL_LAVA : SoundEvents.ITEM_BUCKET_FILL, 1.0F, 1.0F);
+
+                     SoundEvent soundevent = this.fluid.getAttributes().getFillSound();
+                     if (soundevent == null) soundevent = fluid.isIn(FluidTags.LAVA) ? SoundEvents.ITEM_BUCKET_FILL_LAVA : SoundEvents.ITEM_BUCKET_FILL;
+                     playerIn.playSound(soundevent, 1.0F, 1.0F);
                      ItemStack itemstack1 = ItemUsage.method_30012(itemstack, playerIn, new ItemStack(fluid.getBucketItem()));
                      if (!worldIn.isClient) {
                         Criteria.FILLED_BUCKET.trigger((ServerPlayerEntity)playerIn, new ItemStack(fluid.getBucketItem()));
@@ -68,7 +85,7 @@
                return TypedActionResult.fail(itemstack);
             } else {
                BlockState blockstate = worldIn.getBlockState(blockpos);
-               BlockPos blockpos2 = blockstate.getBlock() instanceof FluidFillable && this.fluid == Fluids.WATER ? blockpos : blockpos1;
+               BlockPos blockpos2 = canBlockContainFluid(worldIn, blockpos, blockstate) ? blockpos : blockpos1;
                if (this.placeFluid(playerIn, worldIn, blockpos2, blockraytraceresult)) {
                   this.onEmptied(worldIn, itemstack, blockpos2);
                   if (playerIn instanceof ServerPlayerEntity) {
@@ -116,7 +133,7 @@
             }
 
             return true;
-         } else if (block instanceof FluidFillable && this.fluid == Fluids.WATER) {
+         } else if (block instanceof FluidFillable && ((FluidFillable)block).canFillWithFluid(worldIn,posIn,blockstate,fluid)) {
             ((FluidFillable)block).tryFillWithFluid(worldIn, posIn, blockstate, ((FlowableFluid)this.fluid).getStill(false));
             this.playEmptyingSound(player, worldIn, posIn);
             return true;
@@ -136,7 +153,24 @@
    }
 
    protected void playEmptyingSound(@Nullable PlayerEntity player, WorldAccess worldIn, BlockPos pos) {
-      SoundEvent soundevent = this.fluid.isIn(FluidTags.LAVA) ? SoundEvents.ITEM_BUCKET_EMPTY_LAVA : SoundEvents.ITEM_BUCKET_EMPTY;
+      SoundEvent soundevent = this.fluid.getAttributes().getEmptySound();
+      if(soundevent == null) soundevent = this.fluid.isIn(FluidTags.LAVA) ? SoundEvents.ITEM_BUCKET_EMPTY_LAVA : SoundEvents.ITEM_BUCKET_EMPTY;
       worldIn.playSound(player, pos, soundevent, SoundCategory.BLOCKS, 1.0F, 1.0F);
    }
+
+   @Override
+   public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.CompoundTag nbt) {
+      if (this.getClass() == BucketItem.class)
+         return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+      else
+         return super.initCapabilities(stack, nbt);
+   }
+
+   private final java.util.function.Supplier<? extends Fluid> fluidSupplier;
+   public Fluid getFluid() { return fluidSupplier.get(); }
+
+   private boolean canBlockContainFluid(World worldIn, BlockPos posIn, BlockState blockstate)
+   {
+      return blockstate.getBlock() instanceof FluidFillable && ((FluidFillable)blockstate.getBlock()).canFillWithFluid(worldIn, posIn, blockstate, this.fluid);
+   }
 }
