--- a/net/minecraft/item/DyeableHorseArmorItem.java
+++ b/net/minecraft/item/DyeableHorseArmorItem.java
@@ -4,4 +4,7 @@
    public DyeableHorseArmorItem(int armorValue, String p_i50047_2_, Item.Settings builder) {
       super(armorValue, p_i50047_2_, builder);
    }
+   public DyeableHorseArmorItem(int armorValue, net.minecraft.util.Identifier texture, Item.Settings builder) {
+      super(armorValue, texture, builder);
+   }
 }
