--- a/net/minecraft/item/FilledMapItem.java
+++ b/net/minecraft/item/FilledMapItem.java
@@ -54,6 +54,16 @@
 
    @Nullable
    public static MapState getOrCreateMapState(ItemStack stack, World worldIn) {
+      // FORGE: Add instance method for mods to override
+      Item map = stack.getItem();
+      if (map instanceof FilledMapItem) {
+        return ((FilledMapItem)map).getCustomMapData(stack, worldIn);
+      }
+      return null;
+   }
+
+   @Nullable
+   protected MapState getCustomMapData(ItemStack stack, World worldIn) {
       MapState mapdata = getMapState(stack, worldIn);
       if (mapdata == null && worldIn instanceof ServerWorld) {
          mapdata = createMapState(stack, worldIn, worldIn.getLevelProperties().getSpawnX(), worldIn.getLevelProperties().getSpawnZ(), 3, false, false, worldIn.getRegistryKey());
