--- a/net/minecraft/item/HorseArmorItem.java
+++ b/net/minecraft/item/HorseArmorItem.java
@@ -8,17 +8,21 @@
 
 public class HorseArmorItem extends Item {
    private final int bonus;
-   private final String entityTexture;
+   private final Identifier texture;
 
    public HorseArmorItem(int armorValue, String tierArmor, Item.Settings builder) {
+      this(armorValue, new Identifier("textures/entity/horse/armor/horse_armor_" + tierArmor + ".png"), builder);
+   }
+
+   public HorseArmorItem(int armorValue, Identifier texture, Item.Settings builder) {
       super(builder);
       this.bonus = armorValue;
-      this.entityTexture = "textures/entity/horse/armor/horse_armor_" + tierArmor + ".png";
+      this.texture = texture;
    }
 
    @Environment(EnvType.CLIENT)
    public Identifier getEntityTexture() {
-      return new Identifier(this.entityTexture);
+      return texture;
    }
 
    public int getBonus() {
