--- a/net/minecraft/item/Items.java
+++ b/net/minecraft/item/Items.java
@@ -14,6 +14,7 @@
 import net.minecraft.util.Rarity;
 import net.minecraft.util.registry.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Items {
    public static final Item AIR = register(Blocks.AIR, new AirBlockItem(Blocks.AIR, new Item.Settings()));
    public static final Item STONE = register(Blocks.STONE, ItemGroup.BUILDING_BLOCKS);
