--- a/net/minecraft/client/gui/screen/ingame/AbstractInventoryScreen.java
+++ b/net/minecraft/client/gui/screen/ingame/AbstractInventoryScreen.java
@@ -36,6 +36,9 @@
          this.x = (this.width - this.backgroundWidth) / 2;
          this.drawStatusEffects = false;
       } else {
+         if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.GuiScreenEvent.PotionShiftEvent(this)))
+            this.x = (this.width - this.backgroundWidth) / 2;
+         else
          this.x = 160 + (this.width - this.backgroundWidth - 200) / 2;
          this.drawStatusEffects = true;
       }
@@ -60,7 +63,7 @@
             j = 132 / (collection.size() - 1);
          }
 
-         Iterable<StatusEffectInstance> iterable = Ordering.natural().sortedCopy(collection);
+         Iterable<StatusEffectInstance> iterable = collection.stream().filter(effectInstance -> effectInstance.shouldRender()).sorted().collect(java.util.stream.Collectors.toList());
          this.drawStatusEffectBackgrounds(matrixStack, i, j, iterable);
          this.drawStatusEffectSprites(matrixStack, i, j, iterable);
          this.drawStatusEffectDescriptions(matrixStack, i, j, iterable);
@@ -97,6 +100,11 @@
       int i = this.y;
 
       for(StatusEffectInstance effectinstance : effects) {
+         effectinstance.renderInventoryEffect(this, matrixStack, p_238813_2_, i, this.getZOffset());
+         if (!effectinstance.shouldRenderInvText()) {
+            i += p_238813_3_;
+            continue;
+         }
          String s = I18n.translate(effectinstance.getEffectType().getTranslationKey());
          if (effectinstance.getAmplifier() >= 1 && effectinstance.getAmplifier() <= 9) {
             s = s + ' ' + I18n.translate("enchantment.level." + (effectinstance.getAmplifier() + 1));
