--- a/net/minecraft/entity/ai/goal/BreakDoorGoal.java
+++ b/net/minecraft/entity/ai/goal/BreakDoorGoal.java
@@ -29,7 +29,7 @@
    public boolean canStart() {
       if (!super.canStart()) {
          return false;
-      } else if (!this.mob.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+      } else if (!net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.mob.world, this.doorPos, this.mob)) {
          return false;
       } else {
          return this.isDifficultySufficient(this.mob.world.getDifficulty()) && !this.isDoorOpen();
