--- a/net/minecraft/entity/ai/goal/EatGrassGoal.java
+++ b/net/minecraft/entity/ai/goal/EatGrassGoal.java
@@ -59,7 +59,7 @@
       if (this.timer == 4) {
          BlockPos blockpos = this.mob.getBlockPos();
          if (GRASS_PREDICATE.test(this.world.getBlockState(blockpos))) {
-            if (this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.mob)) {
                this.world.breakBlock(blockpos, false);
             }
 
@@ -67,7 +67,7 @@
          } else {
             BlockPos blockpos1 = blockpos.down();
             if (this.world.getBlockState(blockpos1).isOf(Blocks.GRASS_BLOCK)) {
-               if (this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+               if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.mob)) {
                   this.world.syncWorldEvent(2001, blockpos1, Block.getRawIdFromState(Blocks.GRASS_BLOCK.getDefaultState()));
                   this.world.setBlockState(blockpos1, Blocks.DIRT.getDefaultState(), 2);
                }
