--- a/net/minecraft/util/registry/RegistryKey.java
+++ b/net/minecraft/util/registry/RegistryKey.java
@@ -7,7 +7,7 @@
 import net.minecraft.util.Identifier;
 import net.minecraft.util.registry.Registry;
 
-public class RegistryKey<T> {
+public class RegistryKey<T> implements Comparable<RegistryKey<?>> {
    private static final Map<String, RegistryKey<?>> INSTANCES = Collections.synchronizedMap(Maps.newIdentityHashMap());
    private final Identifier registry;
    private final Identifier value;
@@ -49,4 +49,15 @@
          return of(parent, p_240907_1_);
       };
    }
+
+   public Identifier getRegistryName() {
+      return this.registry;
+   }
+
+   @Override
+   public int compareTo(RegistryKey<?> o) {
+      int ret = this.getRegistryName().compareTo(o.getRegistryName());
+      if (ret == 0) ret = this.getValue().compareTo(o.getValue());
+      return ret;
+   }
 }
