--- a/net/minecraft/advancement/AdvancementRewards.java
+++ b/net/minecraft/advancement/AdvancementRewards.java
@@ -39,7 +39,7 @@
 
    public void apply(ServerPlayerEntity player) {
       player.addExperience(this.experience);
-      LootContext lootcontext = (new LootContext.Builder(player.getServerWorld())).parameter(LootContextParameters.THIS_ENTITY, player).parameter(LootContextParameters.ORIGIN, player.getPos()).random(player.getRandom()).build(LootContextTypes.ADVANCEMENT_REWARD);
+      LootContext lootcontext = (new LootContext.Builder(player.getServerWorld())).parameter(LootContextParameters.THIS_ENTITY, player).parameter(LootContextParameters.ORIGIN, player.getPos()).random(player.getRandom()).luck(player.getLuck()).build(LootContextTypes.ADVANCEMENT_REWARD); // FORGE: luck to LootContext
       boolean flag = false;
 
       for(Identifier resourcelocation : this.loot) {
