--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -58,9 +58,10 @@
       this.setPvpEnabled(true);
       this.setFlightEnabled(true);
       this.method_31400();
+      if (!net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
       this.loadWorld();
       this.setMotd(this.getUserName() + " - " + this.getSaveProperties().getLevelName());
-      return true;
+      return net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void tick(BooleanSupplier hasTimeLeft) {
@@ -174,6 +175,7 @@
    }
 
    public void stop(boolean waitForServer) {
+      if (isRunning())
       this.submitAndJoin(() -> {
          for(ServerPlayerEntity serverplayerentity : Lists.newArrayList(this.getPlayerManager().getPlayerList())) {
             if (!serverplayerentity.getUuid().equals(this.localPlayerUuid)) {
