--- a/net/minecraft/enchantment/FrostWalkerEnchantment.java
+++ b/net/minecraft/enchantment/FrostWalkerEnchantment.java
@@ -42,9 +42,10 @@
             if (blockpos.isWithinDistance(living.getPos(), (double)f)) {
                blockpos$mutable.set(blockpos.getX(), blockpos.getY() + 1, blockpos.getZ());
                BlockState blockstate1 = worldIn.getBlockState(blockpos$mutable);
-               if (blockstate1.isAir()) {
+               if (blockstate1.isAir(worldIn, blockpos$mutable)) {
                   BlockState blockstate2 = worldIn.getBlockState(blockpos);
-                  if (blockstate2.getMaterial() == Material.WATER && blockstate2.get(FluidBlock.LEVEL) == 0 && blockstate.canPlaceAt(worldIn, blockpos) && worldIn.canPlace(blockstate, blockpos, ShapeContext.absent())) {
+                  boolean isFull = blockstate2.getBlock() == Blocks.WATER && blockstate2.get(FluidBlock.LEVEL) == 0; //TODO: Forge, modded waters?
+                  if (blockstate2.getMaterial() == Material.WATER && isFull && blockstate.canPlaceAt(worldIn, blockpos) && worldIn.canPlace(blockstate, blockpos, ShapeContext.absent()) && !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(living, net.minecraftforge.common.util.BlockSnapshot.create(worldIn.getRegistryKey(), worldIn, blockpos), net.minecraft.util.math.Direction.UP)) {
                      worldIn.setBlockState(blockpos, blockstate);
                      worldIn.getBlockTickScheduler().schedule(blockpos, Blocks.FROSTED_ICE, MathHelper.nextInt(living.getRandom(), 60, 120));
                   }
