--- a/net/minecraft/world/ChunkRegion.java
+++ b/net/minecraft/world/ChunkRegion.java
@@ -9,7 +9,6 @@
 import net.fabricmc.api.EnvType;
 import net.fabricmc.api.Environment;
 import net.minecraft.block.Block;
-import net.minecraft.block.BlockEntityProvider;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.entity.BlockEntity;
@@ -170,11 +169,11 @@
 
    public boolean breakBlock(BlockPos pos, boolean dropBlock, @Nullable Entity entity, int recursionLeft) {
       BlockState blockstate = this.getBlockState(pos);
-      if (blockstate.isAir()) {
+      if (blockstate.isAir(this, pos)) {
          return false;
       } else {
          if (dropBlock) {
-            BlockEntity tileentity = blockstate.getBlock().hasBlockEntity() ? this.getBlockEntity(pos) : null;
+            BlockEntity tileentity = blockstate.hasTileEntity() ? this.getBlockEntity(pos) : null;
             Block.dropStacks(blockstate, this.world, pos, tileentity, entity, ItemStack.EMPTY);
          }
 
@@ -194,11 +193,11 @@
          if (compoundnbt != null) {
             if ("DUMMY".equals(compoundnbt.getString("id"))) {
                Block block = blockstate.getBlock();
-               if (!(block instanceof BlockEntityProvider)) {
+               if (!blockstate.hasTileEntity()) {
                   return null;
                }
 
-               tileentity = ((BlockEntityProvider)block).createBlockEntity(this.world);
+               tileentity = blockstate.createTileEntity(this.world);
             } else {
                tileentity = BlockEntity.createFromTag(blockstate, compoundnbt);
             }
@@ -209,7 +208,7 @@
             }
          }
 
-         if (blockstate.getBlock() instanceof BlockEntityProvider) {
+         if (blockstate.hasTileEntity()) {
             LOGGER.warn("Tried to access a block entity before it was created. {}", (Object)pos);
          }
 
@@ -225,9 +224,9 @@
       }
 
       Block block = state.getBlock();
-      if (block.hasBlockEntity()) {
+      if (state.hasTileEntity()) {
          if (ichunk.getStatus().getChunkType() == ChunkStatus.ChunkType.LEVELCHUNK) {
-            ichunk.setBlockEntity(pos, ((BlockEntityProvider)block).createBlockEntity(this));
+            ichunk.setBlockEntity(pos, state.createTileEntity(this));
          } else {
             CompoundTag compoundnbt = new CompoundTag();
             compoundnbt.putInt("x", pos.getX());
@@ -236,7 +235,7 @@
             compoundnbt.putString("id", "DUMMY");
             ichunk.addPendingBlockEntityTag(compoundnbt);
          }
-      } else if (blockstate != null && blockstate.getBlock().hasBlockEntity()) {
+      } else if (blockstate != null && blockstate.hasTileEntity()) {
          ichunk.removeBlockEntity(pos);
       }
 
