--- a/net/minecraft/world/MobSpawnerLogic.java
+++ b/net/minecraft/world/MobSpawnerLogic.java
@@ -132,11 +132,12 @@
                      entity.refreshPositionAndAngles(entity.getX(), entity.getY(), entity.getZ(), world.random.nextFloat() * 360.0F, 0.0F);
                      if (entity instanceof MobEntity) {
                         MobEntity mobentity = (MobEntity)entity;
-                        if (!mobentity.canSpawn(world, SpawnReason.SPAWNER) || !mobentity.canSpawn(world)) {
+                        if (!net.minecraftforge.event.ForgeEventFactory.canEntitySpawnSpawner(mobentity, world, (float)entity.getX(), (float)entity.getY(), (float)entity.getZ(), this)) {
                            continue;
                         }
 
                         if (this.spawnEntry.getEntityTag().getSize() == 1 && this.spawnEntry.getEntityTag().contains("id", 8)) {
+                        if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(mobentity, world, (float)entity.getX(), (float)entity.getY(), (float)entity.getZ(), this, SpawnReason.SPAWNER))
                            ((MobEntity)entity).initialize(serverworld, world.getLocalDifficulty(entity.getBlockPos()), SpawnReason.SPAWNER, (EntityData)null, (CompoundTag)null);
                         }
                      }
@@ -284,4 +285,9 @@
    public double method_8279() {
       return this.field_9159;
    }
+
+   @Nullable
+   public Entity getSpawnerEntity() {
+      return null;
+   }
 }
