--- a/net/minecraft/block/AbstractFireBlock.java
+++ b/net/minecraft/block/AbstractFireBlock.java
@@ -126,6 +126,7 @@
       if (!oldState.isOf(state.getBlock())) {
          if (method_30366(worldIn)) {
             Optional<AreaHelper> optional = AreaHelper.method_30485(worldIn, pos, Direction.Axis.X);
+            optional =  net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, optional);
             if (optional.isPresent()) {
                optional.get().createPortal();
                return;
