--- a/net/minecraft/block/BambooBlock.java
+++ b/net/minecraft/block/BambooBlock.java
@@ -99,10 +99,11 @@
 
    public void randomTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random random) {
       if (state.get(STAGE) == 0) {
-         if (random.nextInt(3) == 0 && worldIn.isAir(pos.up()) && worldIn.getBaseLightLevel(pos.up(), 0) >= 9) {
+         if (worldIn.isAir(pos.up()) && worldIn.getBaseLightLevel(pos.up(), 0) >= 9) {
             int i = this.countBambooBelow(worldIn, pos) + 1;
-            if (i < 16) {
+            if (i < 16 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, random.nextInt(3) == 0)) {
                this.updateLeaves(state, worldIn, pos, random, i);
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
             }
          }
 
