--- a/net/minecraft/block/Blocks.java
+++ b/net/minecraft/block/Blocks.java
@@ -25,6 +25,7 @@
 import net.minecraft.world.BlockView;
 import net.minecraft.world.gen.feature.ConfiguredFeatures;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Blocks {
    public static final Block AIR = register("air", new AirBlock(AbstractBlock.Settings.of(Material.AIR).noCollision().dropsNothing().air()));
    public static final Block STONE = register("stone", new Block(AbstractBlock.Settings.of(Material.STONE, MaterialColor.STONE).requiresTool().strength(1.5F, 6.0F)));
@@ -119,7 +120,7 @@
    public static final Block GREEN_BED = register("green_bed", createBedBlock(DyeColor.GREEN));
    public static final Block RED_BED = register("red_bed", createBedBlock(DyeColor.RED));
    public static final Block BLACK_BED = register("black_bed", createBedBlock(DyeColor.BLACK));
-   public static final Block POWERED_RAIL = register("powered_rail", new PoweredRailBlock(AbstractBlock.Settings.of(Material.SUPPORTED).noCollision().strength(0.7F).sounds(BlockSoundGroup.METAL)));
+   public static final Block POWERED_RAIL = register("powered_rail", new PoweredRailBlock(AbstractBlock.Settings.of(Material.SUPPORTED).noCollision().strength(0.7F).sounds(BlockSoundGroup.METAL), true));
    public static final Block DETECTOR_RAIL = register("detector_rail", new DetectorRailBlock(AbstractBlock.Settings.of(Material.SUPPORTED).noCollision().strength(0.7F).sounds(BlockSoundGroup.METAL)));
    public static final Block STICKY_PISTON = register("sticky_piston", createPistonBlock(true));
    public static final Block COBWEB = register("cobweb", new CobwebBlock(AbstractBlock.Settings.of(Material.COBWEB).noCollision().requiresTool().strength(4.0F)));
@@ -937,15 +938,4 @@
    public static void refreshShapeCache() {
       Block.STATE_IDS.forEach(AbstractBlock.AbstractBlockState::initShapeCache);
    }
-
-   static {
-      for(Block block : Registry.BLOCK) {
-         for(BlockState blockstate : block.getStateManager().getStates()) {
-            Block.STATE_IDS.add(blockstate);
-         }
-
-         block.getLootTableId();
-      }
-
-   }
 }
