--- a/net/minecraft/block/BlockState.java
+++ b/net/minecraft/block/BlockState.java
@@ -6,7 +6,7 @@
 import net.minecraft.state.property.Property;
 import net.minecraft.util.registry.Registry;
 
-public class BlockState extends AbstractBlock.AbstractBlockState {
+public class BlockState extends AbstractBlock.AbstractBlockState implements net.minecraftforge.common.extensions.IForgeBlockState {
    public static final Codec<BlockState> CODEC = createCodec(Registry.BLOCK, Block::getDefaultState).stable();
 
    public BlockState(Block block, ImmutableMap<Property<?>, Comparable<?>> propertiesToValueMap, MapCodec<BlockState> codec) {
