--- a/net/minecraft/block/CampfireBlock.java
+++ b/net/minecraft/block/CampfireBlock.java
@@ -182,7 +182,7 @@
    public void onProjectileHit(World worldIn, BlockState state, BlockHitResult hit, ProjectileEntity projectile) {
       if (!worldIn.isClient && projectile.isOnFire()) {
          Entity entity = projectile.getOwner();
-         boolean flag = entity == null || entity instanceof PlayerEntity || worldIn.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING);
+         boolean flag = entity == null || entity instanceof PlayerEntity || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(worldIn, entity);
          if (flag && !state.get(LIT) && !state.get(WATERLOGGED)) {
             BlockPos blockpos = hit.getBlockPos();
             worldIn.setBlockState(blockpos, state.with(Properties.LIT, Boolean.valueOf(true)), 11);
@@ -209,7 +209,7 @@
             return true;
          }
 
-         boolean flag = VoxelShapes.matchesAnywhere(SMOKEY_SHAPE, blockstate.getCollisionShape(world, pos, ShapeContext.absent()), BooleanBiFunction.AND);
+         boolean flag = VoxelShapes.matchesAnywhere(SMOKEY_SHAPE, blockstate.getCollisionShape(world, blockpos, ShapeContext.absent()), BooleanBiFunction.AND);//Forge fix: MC-201374
          if (flag) {
             BlockState blockstate1 = world.getBlockState(blockpos.down());
             return isLitCampfire(blockstate1);
