--- a/net/minecraft/block/ChorusFlowerBlock.java
+++ b/net/minecraft/block/ChorusFlowerBlock.java
@@ -40,7 +40,7 @@
       BlockPos blockpos = pos.up();
       if (worldIn.isAir(blockpos) && blockpos.getY() < 256) {
          int i = state.get(AGE);
-         if (i < 5) {
+         if (i < 5 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, true)) {
             boolean flag = false;
             boolean flag1 = false;
             BlockState blockstate = worldIn.getBlockState(pos.down());
@@ -65,7 +65,7 @@
                if (j < 2 || j <= random.nextInt(flag1 ? 5 : 4)) {
                   flag = true;
                }
-            } else if (blockstate.isAir()) {
+            } else if (blockstate.isAir(worldIn, pos.down())) {
                flag = true;
             }
 
@@ -97,7 +97,7 @@
             } else {
                this.die(worldIn, pos);
             }
-
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
          }
       }
    }
@@ -133,7 +133,7 @@
    public boolean canPlaceAt(BlockState state, WorldView worldIn, BlockPos pos) {
       BlockState blockstate = worldIn.getBlockState(pos.down());
       if (blockstate.getBlock() != this.plantBlock && !blockstate.isOf(Blocks.END_STONE)) {
-         if (!blockstate.isAir()) {
+         if (!blockstate.isAir(worldIn, pos.down())) {
             return false;
          } else {
             boolean flag = false;
@@ -146,7 +146,7 @@
                   }
 
                   flag = true;
-               } else if (!blockstate1.isAir()) {
+               } else if (!blockstate1.isAir(worldIn, pos.offset(direction))) {
                   return false;
                }
             }
