--- a/net/minecraft/block/CocoaBlock.java
+++ b/net/minecraft/block/CocoaBlock.java
@@ -34,10 +34,11 @@
    }
 
    public void randomTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random random) {
-      if (worldIn.random.nextInt(5) == 0) {
+      if (true) {
          int i = state.get(AGE);
-         if (i < 2) {
+         if (i < 2 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, worldIn.random.nextInt(5) == 0)) {
             worldIn.setBlockState(pos, state.with(AGE, Integer.valueOf(i + 1)), 2);
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
          }
       }
 
