--- a/net/minecraft/block/ComparatorBlock.java
+++ b/net/minecraft/block/ComparatorBlock.java
@@ -154,4 +154,16 @@
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
       builder.add(FACING, MODE, POWERED);
    }
+
+   @Override
+   public boolean getWeakChanges(BlockState state, net.minecraft.world.WorldView world, BlockPos pos) {
+      return state.isOf(Blocks.COMPARATOR);
+   }
+
+   @Override
+   public void onNeighborChange(BlockState state, net.minecraft.world.WorldView world, BlockPos pos, BlockPos neighbor) {
+      if (pos.getY() == neighbor.getY() && world instanceof World && !((World)world).isClient()) {
+         state.neighborUpdate((World)world, pos, world.getBlockState(neighbor).getBlock(), neighbor, false);
+      }
+   }
 }
