--- a/net/minecraft/block/DetectorRailBlock.java
+++ b/net/minecraft/block/DetectorRailBlock.java
@@ -131,6 +131,8 @@
          }
 
          List<AbstractMinecartEntity> list1 = this.getCarts(worldIn, pos, AbstractMinecartEntity.class, EntityPredicates.VALID_INVENTORIES);
+         List<AbstractMinecartEntity> carts = this.getCarts(worldIn, pos, AbstractMinecartEntity.class, null);
+         if (!carts.isEmpty() && carts.get(0).getComparatorLevel() > -1) return carts.get(0).getComparatorLevel();
          if (!list1.isEmpty()) {
             return ScreenHandler.calculateComparatorOutput((Inventory)list1.get(0));
          }
