--- a/net/minecraft/block/DropperBlock.java
+++ b/net/minecraft/block/DropperBlock.java
@@ -37,7 +37,7 @@
          worldIn.syncWorldEvent(1001, pos, 0);
       } else {
          ItemStack itemstack = dispensertileentity.getStack(i);
-         if (!itemstack.isEmpty()) {
+         if (!itemstack.isEmpty() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(worldIn, pos, dispensertileentity, i, itemstack)) {
             Direction direction = worldIn.getBlockState(pos).get(FACING);
             Inventory iinventory = HopperBlockEntity.getInventoryAt(worldIn, pos.offset(direction));
             ItemStack itemstack1;
