--- a/net/minecraft/block/EnchantingTableBlock.java
+++ b/net/minecraft/block/EnchantingTableBlock.java
@@ -55,7 +55,7 @@
             if (rand.nextInt(16) == 0) {
                for(int k = 0; k <= 1; ++k) {
                   BlockPos blockpos = pos.add(i, k, j);
-                  if (worldIn.getBlockState(blockpos).isOf(Blocks.BOOKSHELF)) {
+                  if (worldIn.getBlockState(blockpos).getEnchantPowerBonus(worldIn, blockpos) > 0) {
                      if (!worldIn.isAir(pos.add(i / 2, 0, j / 2))) {
                         break;
                      }
