--- a/net/minecraft/block/FallingBlock.java
+++ b/net/minecraft/block/FallingBlock.java
@@ -31,7 +31,7 @@
    }
 
    public void scheduledTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random rand) {
-      if (canFallThrough(worldIn.getBlockState(pos.down())) && pos.getY() >= 0) {
+      if (worldIn.isAir(pos.down()) || canFallThrough(worldIn.getBlockState(pos.down())) && pos.getY() >= 0) {
          FallingBlockEntity fallingblockentity = new FallingBlockEntity(worldIn, (double)pos.getX() + 0.5D, (double)pos.getY(), (double)pos.getZ() + 0.5D, worldIn.getBlockState(pos));
          this.configureFallingBlockEntity(fallingblockentity);
          worldIn.spawnEntity(fallingblockentity);
@@ -60,7 +60,7 @@
    public void randomDisplayTick(BlockState stateIn, World worldIn, BlockPos pos, Random rand) {
       if (rand.nextInt(16) == 0) {
          BlockPos blockpos = pos.down();
-         if (canFallThrough(worldIn.getBlockState(blockpos))) {
+         if (worldIn.isAir(blockpos) || canFallThrough(worldIn.getBlockState(blockpos))) {
             double d0 = (double)pos.getX() + rand.nextDouble();
             double d1 = (double)pos.getY() - 0.05D;
             double d2 = (double)pos.getZ() + rand.nextDouble();
