--- a/net/minecraft/block/FarmlandBlock.java
+++ b/net/minecraft/block/FarmlandBlock.java
@@ -76,7 +76,7 @@
    }
 
    public void onLandedUpon(World worldIn, BlockPos pos, Entity entityIn, float fallDistance) {
-      if (!worldIn.isClient && worldIn.random.nextFloat() < fallDistance - 0.5F && entityIn instanceof LivingEntity && (entityIn instanceof PlayerEntity || worldIn.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) && entityIn.getWidth() * entityIn.getWidth() * entityIn.getHeight() > 0.512F) {
+      if (!worldIn.isClient && net.minecraftforge.common.ForgeHooks.onFarmlandTrample(worldIn, pos, Blocks.DIRT.getDefaultState(), fallDistance, entityIn)) { // Forge: Move logic to Entity#canTrample
          setToDirt(worldIn.getBlockState(pos), worldIn, pos);
       }
 
@@ -87,9 +87,10 @@
       worldIn.setBlockState(pos, pushEntitiesUpBeforeBlockChange(state, Blocks.DIRT.getDefaultState(), worldIn, pos));
    }
 
-   private static boolean hasCrop(BlockView worldIn, BlockPos pos) {
-      Block block = worldIn.getBlockState(pos.up()).getBlock();
-      return block instanceof CropBlock || block instanceof StemBlock || block instanceof AttachedStemBlock;
+   private boolean hasCrop(BlockView worldIn, BlockPos pos) {
+      BlockState plant = worldIn.getBlockState(pos.up());
+      BlockState state = worldIn.getBlockState(pos);
+      return plant.getBlock() instanceof net.minecraftforge.common.IPlantable && state.canSustainPlant(worldIn, pos, Direction.UP, (net.minecraftforge.common.IPlantable)plant.getBlock());
    }
 
    private static boolean isWaterNearby(WorldView worldIn, BlockPos pos) {
@@ -99,7 +100,7 @@
          }
       }
 
-      return false;
+      return net.minecraftforge.common.FarmlandWaterManager.hasBlockWaterTicket(worldIn, pos);
    }
 
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
