--- a/net/minecraft/block/NetherWartBlock.java
+++ b/net/minecraft/block/NetherWartBlock.java
@@ -34,9 +34,10 @@
 
    public void randomTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random random) {
       int i = state.get(AGE);
-      if (i < 3 && random.nextInt(10) == 0) {
+      if (i < 3 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, random.nextInt(10) == 0)) {
          state = state.with(AGE, Integer.valueOf(i + 1));
          worldIn.setBlockState(pos, state, 2);
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state);
       }
 
    }
