--- a/net/minecraft/block/NoteBlock.java
+++ b/net/minecraft/block/NoteBlock.java
@@ -50,7 +50,7 @@
    }
 
    private void playNote(World worldIn, BlockPos pos) {
-      if (worldIn.getBlockState(pos.up()).isAir()) {
+      if (worldIn.isAir(pos.up())) {
          worldIn.addSyncedBlockEvent(pos, this, 0, 0);
       }
 
@@ -60,7 +60,9 @@
       if (worldIn.isClient) {
          return ActionResult.SUCCESS;
       } else {
-         state = state.cycle(NOTE);
+         int _new = net.minecraftforge.common.ForgeHooks.onNoteChange(worldIn, pos, state, state.get(NOTE), state.cycle(NOTE).get(NOTE));
+         if (_new == -1) return ActionResult.FAIL;
+         state = state.with(NOTE, _new);
          worldIn.setBlockState(pos, state, 3);
          this.playNote(worldIn, pos);
          player.incrementStat(Stats.TUNE_NOTEBLOCK);
@@ -76,6 +78,9 @@
    }
 
    public boolean onSyncedBlockEvent(BlockState state, World worldIn, BlockPos pos, int id, int param) {
+      net.minecraftforge.event.world.NoteBlockEvent.Play e = new net.minecraftforge.event.world.NoteBlockEvent.Play(worldIn, pos, state, state.get(NOTE), state.get(INSTRUMENT));
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e)) return false;
+      state = state.with(NOTE, e.getVanillaNoteId()).with(INSTRUMENT, e.getInstrument());
       int i = state.get(NOTE);
       float f = (float)Math.pow(2.0D, (double)(i - 12) / 12.0D);
       worldIn.playSound((PlayerEntity)null, pos, state.get(INSTRUMENT).getSound(), SoundCategory.RECORDS, 3.0F, f);
