--- a/net/minecraft/block/OreBlock.java
+++ b/net/minecraft/block/OreBlock.java
@@ -31,12 +31,10 @@
 
    public void onStacksDropped(BlockState state, ServerWorld worldIn, BlockPos pos, ItemStack stack) {
       super.onStacksDropped(state, worldIn, pos, stack);
-      if (EnchantmentHelper.getLevel(Enchantments.SILK_TOUCH, stack) == 0) {
-         int i = this.getExperienceWhenMined(worldIn.random);
-         if (i > 0) {
-            this.dropExperience(worldIn, pos, i);
-         }
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.WorldView reader, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? this.getExperienceWhenMined(RANDOM) : 0;
    }
 }
