--- a/net/minecraft/block/PistonBlock.java
+++ b/net/minecraft/block/PistonBlock.java
@@ -158,6 +158,7 @@
       }
 
       if (id == 0) {
+         if (net.minecraftforge.event.ForgeEventFactory.onPistonMovePre(worldIn, pos, direction, true)) return false;
          if (!this.move(worldIn, pos, direction, true)) {
             return false;
          }
@@ -165,6 +166,7 @@
          worldIn.setBlockState(pos, state.with(EXTENDED, Boolean.valueOf(true)), 67);
          worldIn.playSound((PlayerEntity)null, pos, SoundEvents.BLOCK_PISTON_EXTEND, SoundCategory.BLOCKS, 0.5F, worldIn.random.nextFloat() * 0.25F + 0.6F);
       } else if (id == 1 || id == 2) {
+         if (net.minecraftforge.event.ForgeEventFactory.onPistonMovePre(worldIn, pos, direction, false)) return false;
          BlockEntity tileentity1 = worldIn.getBlockEntity(pos.offset(direction));
          if (tileentity1 instanceof PistonBlockEntity) {
             ((PistonBlockEntity)tileentity1).finish();
@@ -204,6 +206,7 @@
          worldIn.playSound((PlayerEntity)null, pos, SoundEvents.BLOCK_PISTON_CONTRACT, SoundCategory.BLOCKS, 0.5F, worldIn.random.nextFloat() * 0.15F + 0.6F);
       }
 
+      net.minecraftforge.event.ForgeEventFactory.onPistonMovePost(worldIn, pos, direction, (id == 0));
       return true;
    }
 
@@ -234,7 +237,7 @@
                   return false;
                }
 
-               return !blockStateIn.getBlock().hasBlockEntity();
+               return !blockStateIn.hasTileEntity();
             }
          } else {
             return false;
@@ -273,7 +276,7 @@
          for(int k = list2.size() - 1; k >= 0; --k) {
             BlockPos blockpos2 = list2.get(k);
             BlockState blockstate1 = worldIn.getBlockState(blockpos2);
-            BlockEntity tileentity = blockstate1.getBlock().hasBlockEntity() ? worldIn.getBlockEntity(blockpos2) : null;
+            BlockEntity tileentity = blockstate1.hasTileEntity() ? worldIn.getBlockEntity(blockpos2) : null;
             dropStacks(blockstate1, worldIn, blockpos2, tileentity);
             worldIn.setBlockState(blockpos2, Blocks.AIR.getDefaultState(), 18);
             ablockstate[j++] = blockstate1;
@@ -337,6 +340,10 @@
       return state.with(FACING, rot.rotate(state.get(FACING)));
    }
 
+   public BlockState rotate(BlockState state, net.minecraft.world.WorldAccess world, BlockPos pos, BlockRotation direction) {
+       return state.get(EXTENDED) ? state : super.rotate(state, world, pos, direction);
+   }
+
    public BlockState mirror(BlockState state, BlockMirror mirrorIn) {
       return state.rotate(mirrorIn.getRotation(state.get(FACING)));
    }
