--- a/net/minecraft/block/RailBlock.java
+++ b/net/minecraft/block/RailBlock.java
@@ -49,6 +49,9 @@
             return state.with(SHAPE, RailShape.SOUTH_EAST);
          case NORTH_EAST:
             return state.with(SHAPE, RailShape.SOUTH_WEST);
+         case NORTH_SOUTH: //Forge fix: MC-196102
+         case EAST_WEST:
+            return state;
          }
       case COUNTERCLOCKWISE_90:
          switch((RailShape)state.get(SHAPE)) {
