--- a/net/minecraft/block/RedstoneWireBlock.java
+++ b/net/minecraft/block/RedstoneWireBlock.java
@@ -185,7 +185,7 @@
       BlockState blockstate = reader.getBlockState(blockpos);
       if (nonNormalCubeAbove) {
          boolean flag = this.canRunOnTop(reader, blockpos, blockstate);
-         if (flag && connectsTo(reader.getBlockState(blockpos.up()))) {
+         if (flag && canConnectTo(reader.getBlockState(blockpos.up()), reader, blockpos.up(), null) ) {
             if (blockstate.isSideSolidFullSquare(reader, blockpos, direction.getOpposite())) {
                return WireConnection.UP;
             }
@@ -194,7 +194,7 @@
          }
       }
 
-      return !connectsTo(blockstate, direction) && (blockstate.isSolidBlock(reader, blockpos) || !connectsTo(reader.getBlockState(blockpos.down()))) ? WireConnection.NONE : WireConnection.SIDE;
+      return !canConnectTo(blockstate, reader, blockpos, direction) && (blockstate.isSolidBlock(reader, blockpos) || !canConnectTo(reader.getBlockState(blockpos.down()), reader, blockpos.down(), null)) ? WireConnection.NONE : WireConnection.SIDE;
    }
 
    public boolean canPlaceAt(BlockState state, WorldView worldIn, BlockPos pos) {
@@ -336,11 +336,7 @@
       }
    }
 
-   protected static boolean connectsTo(BlockState state) {
-      return connectsTo(state, (Direction)null);
-   }
-
-   protected static boolean connectsTo(BlockState blockState, @Nullable Direction side) {
+   protected static boolean canConnectTo(BlockState blockState, BlockView world, BlockPos pos, @Nullable Direction side) {
       if (blockState.isOf(Blocks.REDSTONE_WIRE)) {
          return true;
       } else if (blockState.isOf(Blocks.REPEATER)) {
@@ -349,7 +345,7 @@
       } else if (blockState.isOf(Blocks.OBSERVER)) {
          return side == blockState.get(ObserverBlock.FACING);
       } else {
-         return blockState.emitsRedstonePower() && side != null;
+         return blockState.canConnectRedstone(world, pos, side) && side != null;
       }
    }
 
