--- a/net/minecraft/block/SaplingBlock.java
+++ b/net/minecraft/block/SaplingBlock.java
@@ -28,6 +28,7 @@
 
    public void randomTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random random) {
       if (worldIn.getLightLevel(pos.up()) >= 9 && random.nextInt(7) == 0) {
+      if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
          this.generate(worldIn, pos, state, random);
       }
 
@@ -37,6 +38,7 @@
       if (state.get(STAGE) == 0) {
          world.setBlockState(pos, state.cycle(STAGE), 4);
       } else {
+         if (!net.minecraftforge.event.ForgeEventFactory.saplingGrowTree(world, rand, pos)) return;
          this.generator.generate(world, world.getChunkManager().getChunkGenerator(), pos, state, rand);
       }
 
