--- a/net/minecraft/block/SpawnerBlock.java
+++ b/net/minecraft/block/SpawnerBlock.java
@@ -18,10 +18,13 @@
 
    public void onStacksDropped(BlockState state, ServerWorld worldIn, BlockPos pos, ItemStack stack) {
       super.onStacksDropped(state, worldIn, pos, stack);
-      int i = 15 + worldIn.random.nextInt(15) + worldIn.random.nextInt(15);
-      this.dropExperience(worldIn, pos, i);
    }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.WorldView world, BlockPos pos, int fortune, int silktouch) {
+      return 15 + RANDOM.nextInt(15) + RANDOM.nextInt(15);
+   }
+
    public BlockRenderType getRenderType(BlockState state) {
       return BlockRenderType.MODEL;
    }
