--- a/net/minecraft/block/SpongeBlock.java
+++ b/net/minecraft/block/SpongeBlock.java
@@ -63,7 +63,7 @@
                      queue.add(new Pair<>(blockpos1, j + 1));
                   }
                } else if (material == Material.UNDERWATER_PLANT || material == Material.REPLACEABLE_UNDERWATER_PLANT) {
-                  BlockEntity tileentity = blockstate.getBlock().hasBlockEntity() ? worldIn.getBlockEntity(blockpos1) : null;
+                  BlockEntity tileentity = blockstate.hasTileEntity() ? worldIn.getBlockEntity(blockpos1) : null;
                   dropStacks(blockstate, worldIn, blockpos1, tileentity);
                   worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 3);
                   ++i;
