--- a/net/minecraft/block/StairsBlock.java
+++ b/net/minecraft/block/StairsBlock.java
@@ -84,13 +84,23 @@
       return voxelshape;
    }
 
+   @Deprecated // Forge: Use the other constructor that takes a Supplier
    public StairsBlock(BlockState state, AbstractBlock.Settings properties) {
       super(properties);
       this.setDefaultState(this.stateManager.getDefaultState().with(FACING, Direction.NORTH).with(HALF, BlockHalf.BOTTOM).with(SHAPE, StairShape.STRAIGHT).with(WATERLOGGED, Boolean.valueOf(false)));
       this.baseBlock = state.getBlock();
       this.baseBlockState = state;
+      this.stateSupplier = () -> state;
    }
 
+   public StairsBlock(java.util.function.Supplier<BlockState> state, AbstractBlock.Settings properties) {
+      super(properties);
+      this.setDefaultState(this.stateManager.getDefaultState().with(FACING, Direction.NORTH).with(HALF, BlockHalf.BOTTOM).with(SHAPE, StairShape.STRAIGHT).with(WATERLOGGED, Boolean.valueOf(false)));
+      this.baseBlock = Blocks.AIR; // These are unused, fields are redirected
+      this.baseBlockState = Blocks.AIR.getDefaultState();
+      this.stateSupplier = state;
+   }
+
    public boolean hasSidedTransparency(BlockState state) {
       return true;
    }
@@ -266,4 +276,14 @@
    public boolean canPathfindThrough(BlockState state, BlockView worldIn, BlockPos pos, NavigationType type) {
       return false;
    }
+
+   // Forge Start
+   private final java.util.function.Supplier<BlockState> stateSupplier;
+   private Block getModelBlock() {
+       return getModelState().getBlock();
+   }
+   private BlockState getModelState() {
+       return stateSupplier.get();
+   }
+   // Forge end
 }
