--- a/net/minecraft/block/TurtleEggBlock.java
+++ b/net/minecraft/block/TurtleEggBlock.java
@@ -148,7 +148,7 @@
          if (!(trampler instanceof LivingEntity)) {
             return false;
          } else {
-            return trampler instanceof PlayerEntity || worldIn.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING);
+            return trampler instanceof PlayerEntity || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(worldIn, trampler);
          }
       } else {
          return false;
