--- a/net/minecraft/block/VineBlock.java
+++ b/net/minecraft/block/VineBlock.java
@@ -21,7 +21,7 @@
 import net.minecraft.world.WorldAccess;
 import net.minecraft.world.WorldView;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty UP = ConnectingBlock.UP;
    public static final BooleanProperty NORTH = ConnectingBlock.NORTH;
    public static final BooleanProperty EAST = ConnectingBlock.EAST;
@@ -151,14 +151,14 @@
    }
 
    public void randomTick(BlockState state, ServerWorld worldIn, BlockPos pos, Random random) {
-      if (worldIn.random.nextInt(4) == 0) {
+      if (worldIn.random.nextInt(4) == 0 && worldIn.isAreaLoaded(pos, 4)) { // Forge: check area to prevent loading unloaded chunks
          Direction direction = Direction.random(random);
          BlockPos blockpos = pos.up();
          if (direction.getAxis().isHorizontal() && !state.get(getFacingProperty(direction))) {
             if (this.canGrowAt(worldIn, pos)) {
                BlockPos blockpos4 = pos.offset(direction);
                BlockState blockstate4 = worldIn.getBlockState(blockpos4);
-               if (blockstate4.isAir()) {
+               if (blockstate4.isAir(worldIn, blockpos4)) {
                   Direction direction3 = direction.rotateYClockwise();
                   Direction direction4 = direction.rotateYCounterclockwise();
                   boolean flag = state.get(getFacingProperty(direction3));
@@ -215,8 +215,9 @@
             if (pos.getY() > 0) {
                BlockPos blockpos1 = pos.down();
                BlockState blockstate = worldIn.getBlockState(blockpos1);
-               if (blockstate.isAir() || blockstate.isOf(this)) {
-                  BlockState blockstate1 = blockstate.isAir() ? this.getDefaultState() : blockstate;
+               boolean isAir = blockstate.isAir(worldIn, blockpos1);
+               if (isAir || blockstate.isOf(this)) {
+                  BlockState blockstate1 = isAir ? this.getDefaultState() : blockstate;
                   BlockState blockstate2 = this.getGrownState(state, blockstate1, random);
                   if (blockstate1 != blockstate2 && this.hasHorizontalSide(blockstate2)) {
                      worldIn.setBlockState(blockpos1, blockstate2, 2);
@@ -321,4 +322,7 @@
    public static BooleanProperty getFacingProperty(Direction side) {
       return FACING_PROPERTIES.get(side);
    }
+
+   @Override
+   public boolean isLadder(BlockState state, WorldView world, BlockPos pos, net.minecraft.entity.LivingEntity entity) { return true; }
 }
