--- a/net/minecraft/fluid/LavaFluid.java
+++ b/net/minecraft/fluid/LavaFluid.java
@@ -73,7 +73,7 @@
                BlockState blockstate = world.getBlockState(blockpos);
                if (blockstate.isAir()) {
                   if (this.canLightFire(world, blockpos)) {
-                     world.setBlockState(blockpos, AbstractFireBlock.getState(world, blockpos));
+                     world.setBlockState(blockpos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(world, blockpos, pos, Blocks.FIRE.getDefaultState()));
                      return;
                   }
                } else if (blockstate.getMaterial().blocksMovement()) {
@@ -88,7 +88,7 @@
                }
 
                if (world.isAir(blockpos1.up()) && this.hasBurnableBlock(world, blockpos1)) {
-                  world.setBlockState(blockpos1.up(), AbstractFireBlock.getState(world, blockpos1));
+                  world.setBlockState(blockpos1.up(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(world, blockpos1.up(), pos, Blocks.FIRE.getDefaultState()));
                }
             }
          }
@@ -166,7 +166,7 @@
          FluidState fluidstate = worldIn.getFluidState(pos);
          if (this.isIn(FluidTags.LAVA) && fluidstate.isIn(FluidTags.WATER)) {
             if (blockStateIn.getBlock() instanceof FluidBlock) {
-               worldIn.setBlockState(pos, Blocks.STONE.getDefaultState(), 3);
+               worldIn.setBlockState(pos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, pos, pos, Blocks.STONE.getDefaultState()), 3);
             }
 
             this.playExtinguishEvent(worldIn, pos);
