--- a/net/minecraft/fluid/WaterFluid.java
+++ b/net/minecraft/fluid/WaterFluid.java
@@ -62,7 +62,7 @@
    }
 
    protected void beforeBreakingBlock(WorldAccess worldIn, BlockPos pos, BlockState state) {
-      BlockEntity tileentity = state.getBlock().hasBlockEntity() ? worldIn.getBlockEntity(pos) : null;
+      BlockEntity tileentity = state.hasTileEntity() ? worldIn.getBlockEntity(pos) : null;
       Block.dropStacks(state, worldIn, pos, tileentity);
    }
 
