--- a/net/minecraft/world/gen/feature/IceSpikeFeature.java
+++ b/net/minecraft/world/gen/feature/IceSpikeFeature.java
@@ -42,14 +42,14 @@
                   if ((i1 == 0 && j1 == 0 || !(f1 * f1 + f2 * f2 > f * f)) && (i1 != -l && i1 != l && j1 != -l && j1 != l || !(rand.nextFloat() > 0.75F))) {
                      BlockState blockstate = reader.getBlockState(pos.add(i1, k, j1));
                      Block block = blockstate.getBlock();
-                     if (blockstate.isAir() || isSoil(block) || block == Blocks.SNOW_BLOCK || block == Blocks.ICE) {
+                     if (blockstate.isAir(reader, pos.add(i1, k, j1)) || isSoil(block) || block == Blocks.SNOW_BLOCK || block == Blocks.ICE) {
                         this.setBlockState(reader, pos.add(i1, k, j1), Blocks.PACKED_ICE.getDefaultState());
                      }
 
                      if (k != 0 && l > 1) {
                         blockstate = reader.getBlockState(pos.add(i1, -k, j1));
                         block = blockstate.getBlock();
-                        if (blockstate.isAir() || isSoil(block) || block == Blocks.SNOW_BLOCK || block == Blocks.ICE) {
+                        if (blockstate.isAir(reader, pos.add(i1, -k, j1)) || isSoil(block) || block == Blocks.SNOW_BLOCK || block == Blocks.ICE) {
                            this.setBlockState(reader, pos.add(i1, -k, j1), Blocks.PACKED_ICE.getDefaultState());
                         }
                      }
@@ -76,7 +76,7 @@
                while(blockpos.getY() > 50) {
                   BlockState blockstate1 = reader.getBlockState(blockpos);
                   Block block1 = blockstate1.getBlock();
-                  if (!blockstate1.isAir() && !isSoil(block1) && block1 != Blocks.SNOW_BLOCK && block1 != Blocks.ICE && block1 != Blocks.PACKED_ICE) {
+                  if (!blockstate1.isAir(reader, blockpos) && !isSoil(block1) && block1 != Blocks.SNOW_BLOCK && block1 != Blocks.ICE && block1 != Blocks.PACKED_ICE) {
                      break;
                   }
 
