--- a/net/minecraft/client/gl/ShaderEffect.java
+++ b/net/minecraft/client/gl/ShaderEffect.java
@@ -165,7 +165,8 @@
                         throw new ShaderParseException("Render target '" + s4 + "' can't be used as depth buffer");
                      }
 
-                     Identifier resourcelocation = new Identifier("textures/effect/" + s4 + ".png");
+                     Identifier rl = Identifier.tryParse(s4);
+                     Identifier resourcelocation = new Identifier(rl.getNamespace(), "textures/effect/" + rl.getPath() + ".png");
                      Resource iresource = null;
 
                      try {
