--- a/net/minecraft/entity/boss/dragon/EnderDragonEntity.java
+++ b/net/minecraft/entity/boss/dragon/EnderDragonEntity.java
@@ -409,8 +409,8 @@
                BlockPos blockpos = new BlockPos(k1, l1, i2);
                BlockState blockstate = this.world.getBlockState(blockpos);
                Block block = blockstate.getBlock();
-               if (!blockstate.isAir() && blockstate.getMaterial() != Material.FIRE) {
-                  if (this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING) && !BlockTags.DRAGON_IMMUNE.contains(block)) {
+               if (!blockstate.isAir(this.world, blockpos) && blockstate.getMaterial() != Material.FIRE) {
+                  if (net.minecraftforge.common.ForgeHooks.canEntityDestroy(this.world, blockpos, this) && !BlockTags.DRAGON_IMMUNE.contains(block)) {
                      flag1 = this.world.removeBlock(blockpos, false) || flag1;
                   } else {
                      flag = true;
