--- a/net/minecraft/client/world/ClientWorld.java
+++ b/net/minecraft/client/world/ClientWorld.java
@@ -104,6 +104,8 @@
       this.setSpawnPos(new BlockPos(8, 64, 8), 0.0F);
       this.calculateAmbientDarkness();
       this.initWeatherGradients();
+      this.gatherCapabilities();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(this));
    }
 
    public SkyProperties getSkyProperties() {
@@ -186,6 +188,7 @@
             this.getProfiler().push(() -> {
                return Registry.ENTITY_TYPE.getId(entityIn.getType()).toString();
             });
+            if (entityIn.canUpdate())
             entityIn.tick();
             this.getProfiler().pop();
          }
@@ -285,9 +288,11 @@
    }
 
    private void addEntityPrivate(int entityIdIn, Entity entityToSpawn) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(entityToSpawn, this))) return;
       this.removeEntity(entityIdIn);
       this.regularEntities.put(entityIdIn, entityToSpawn);
       this.getChunkManager().getChunk(MathHelper.floor(entityToSpawn.getX() / 16.0D), MathHelper.floor(entityToSpawn.getZ() / 16.0D), ChunkStatus.FULL, true).addEntity(entityToSpawn);
+      entityToSpawn.onAddedToWorld();
    }
 
    public void removeEntity(int eid) {
@@ -306,6 +311,8 @@
       }
 
       this.players.remove(entityIn);
+      entityIn.onRemovedFromWorld();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityLeaveWorldEvent(entityIn, this));
    }
 
    public void addEntitiesToChunk(WorldChunk chunkIn) {
@@ -448,6 +455,11 @@
    }
 
    public void playSound(@Nullable PlayerEntity player, double x, double y, double z, SoundEvent soundIn, SoundCategory category, float volume, float pitch) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(player, soundIn, category, volume, pitch);
+      if (event.isCanceled() || event.getSound() == null) return;
+      soundIn = event.getSound();
+      category = event.getCategory();
+      volume = event.getVolume();
       if (player == this.client.player) {
          this.playSound(x, y, z, soundIn, category, volume, pitch, false);
       }
@@ -455,6 +467,11 @@
    }
 
    public void playSoundFromEntity(@Nullable PlayerEntity playerIn, Entity entityIn, SoundEvent eventIn, SoundCategory categoryIn, float volume, float pitch) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(playerIn, eventIn, categoryIn, volume, pitch);
+      if (event.isCanceled() || event.getSound() == null) return;
+      eventIn = event.getSound();
+      categoryIn = event.getCategory();
+      volume = event.getVolume();
       if (playerIn == this.client.player) {
          this.client.getSoundManager().play(new EntityTrackingSoundInstance(eventIn, categoryIn, entityIn));
       }
@@ -876,6 +893,7 @@
       }
 
       public void setDifficulty(Difficulty difficulty) {
+         net.minecraftforge.common.ForgeHooks.onDifficultyChange(difficulty, this.difficulty);
          this.difficulty = difficulty;
       }
 
