--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -122,7 +122,7 @@
       String s7 = getOption(optionset, optionspec);
       Integer integer = getOption(optionset, optionspec1);
       CrashReport.initCrashReport();
-      Bootstrap.initialize();
+      net.minecraftforge.fml.loading.BackgroundWaiter.runAndTick(()->Bootstrap.initialize(), net.minecraftforge.fml.loading.FMLLoader.progressWindowTick);
       Bootstrap.logMissing();
       Util.startTimerHack();
       Session session = new Session(optionspec10.value(optionset), s5, optionspec12.value(optionset), optionspec21.value(optionset));
