--- a/net/minecraft/world/border/WorldBorder.java
+++ b/net/minecraft/world/border/WorldBorder.java
@@ -138,6 +138,10 @@
       this.listeners.add(listener);
    }
 
+   public void removeListener(WorldBorderListener listener) {
+      this.listeners.remove(listener);
+   }
+
    public void setMaxWorldBorderRadius(int size) {
       this.maxWorldBorderRadius = size;
       this.area.onMaxWorldBorderRadiusChanged();
