--- a/net/minecraft/server/world/ThreadedAnvilChunkStorage.java
+++ b/net/minecraft/server/world/ThreadedAnvilChunkStorage.java
@@ -399,6 +399,7 @@
             if (this.chunksToUnload.remove(chunkPosIn, chunkHolderIn) && p_219185_5_ != null) {
                if (p_219185_5_ instanceof WorldChunk) {
                   ((WorldChunk)p_219185_5_).setLoadedToWorld(false);
+                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload((WorldChunk)p_219185_5_));
                }
 
                this.save(p_219185_5_);
@@ -600,6 +601,7 @@
                if (list != null) {
                   list.forEach(chunk::remove);
                }
+               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
             }
 
             return chunk;
@@ -677,6 +679,7 @@
 
             this.world.getProfiler().visit("chunkSave");
             CompoundTag compoundnbt = ChunkSerializer.serialize(this.world, chunkIn);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkDataEvent.Save(chunkIn, chunkIn.getWorldForge() != null ? chunkIn.getWorldForge() : this.world, compoundnbt));
             this.setTagAt(chunkpos, compoundnbt);
             this.method_27053(chunkpos, chunkstatus.getChunkType());
             return true;
@@ -733,6 +736,7 @@
 
    protected void sendWatchPackets(ServerPlayerEntity player, ChunkPos chunkPosIn, Packet<?>[] packetCache, boolean wasLoaded, boolean load) {
       if (player.world == this.world) {
+         net.minecraftforge.event.ForgeEventFactory.fireChunkWatch(wasLoaded, load, player, chunkPosIn, this.world);
          if (load && !wasLoaded) {
             ChunkHolder chunkholder = this.getChunkHolder(chunkPosIn.toLong());
             if (chunkholder != null) {
