--- a/net/minecraft/loot/condition/RandomChanceWithLootingLootCondition.java
+++ b/net/minecraft/loot/condition/RandomChanceWithLootingLootCondition.java
@@ -32,12 +32,7 @@
    }
 
    public boolean test(LootContext p_test_1_) {
-      Entity entity = p_test_1_.get(LootContextParameters.KILLER_ENTITY);
-      int i = 0;
-      if (entity instanceof LivingEntity) {
-         i = EnchantmentHelper.getLooting((LivingEntity)entity);
-      }
-
+      int i = p_test_1_.getLootingModifier();
       return p_test_1_.getRandom().nextFloat() < this.chance + (float)i * this.lootingMultiplier;
    }
 
