--- a/net/minecraft/recipe/SmithingRecipe.java
+++ b/net/minecraft/recipe/SmithingRecipe.java
@@ -65,7 +65,7 @@
       return RecipeType.SMITHING;
    }
 
-   public static class Serializer implements RecipeSerializer<SmithingRecipe> {
+   public static class Serializer extends net.minecraftforge.registries.ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<SmithingRecipe> {
       public SmithingRecipe read(Identifier recipeId, JsonObject json) {
          Ingredient ingredient = Ingredient.fromJson(JsonHelper.getObject(json, "base"));
          Ingredient ingredient1 = Ingredient.fromJson(JsonHelper.getObject(json, "addition"));
