--- a/net/minecraft/client/particle/Particle.java
+++ b/net/minecraft/client/particle/Particle.java
@@ -210,4 +210,12 @@
    public void setBoundingBox(Box bb) {
       this.boundingBox = bb;
    }
+
+   /**
+    * Forge added method that controls if a particle should be culled to it's bounding box.
+    * Default behaviour is culling enabled
+    */
+   public boolean shouldCull() {
+      return true;
+   }
 }
