--- a/net/minecraft/entity/mob/AbstractSkeletonEntity.java
+++ b/net/minecraft/entity/mob/AbstractSkeletonEntity.java
@@ -154,7 +154,7 @@
          this.goalSelector.remove(this.meleeAttackGoal);
          this.goalSelector.remove(this.bowAttackGoal);
          ItemStack itemstack = this.getStackInHand(ProjectileUtil.getHandPossiblyHolding(this, Items.BOW));
-         if (itemstack.getItem() == Items.BOW) {
+         if (itemstack.getItem() instanceof net.minecraft.item.BowItem) {
             int i = 20;
             if (this.world.getDifficulty() != Difficulty.HARD) {
                i = 40;
@@ -172,6 +172,8 @@
    public void attack(LivingEntity target, float distanceFactor) {
       ItemStack itemstack = this.getArrowType(this.getStackInHand(ProjectileUtil.getHandPossiblyHolding(this, Items.BOW)));
       PersistentProjectileEntity abstractarrowentity = this.createArrowProjectile(itemstack, distanceFactor);
+      if (this.getMainHandStack().getItem() instanceof net.minecraft.item.BowItem)
+         abstractarrowentity = ((net.minecraft.item.BowItem)this.getMainHandStack().getItem()).customArrow(abstractarrowentity);
       double d0 = target.getX() - this.getX();
       double d1 = target.getBodyY(0.3333333333333333D) - abstractarrowentity.getY();
       double d2 = target.getZ() - this.getZ();
