--- a/net/minecraft/entity/mob/CreeperEntity.java
+++ b/net/minecraft/entity/mob/CreeperEntity.java
@@ -217,7 +217,7 @@
 
    private void explode() {
       if (!this.world.isClient) {
-         Explosion.DestructionType explosion$mode = this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
+         Explosion.DestructionType explosion$mode = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this) ? Explosion.DestructionType.DESTROY : Explosion.DestructionType.NONE;
          float f = this.shouldRenderOverlay() ? 2.0F : 1.0F;
          this.dead = true;
          this.world.createExplosion(this, this.getX(), this.getY(), this.getZ(), (float)this.explosionRadius * f, explosion$mode);
