--- a/net/minecraft/world/level/LevelInfo.java
+++ b/net/minecraft/world/level/LevelInfo.java
@@ -65,6 +65,7 @@
    }
 
    public LevelInfo withDifficulty(Difficulty difficulty) {
+      net.minecraftforge.common.ForgeHooks.onDifficultyChange(difficulty, this.difficulty);
       return new LevelInfo(this.name, this.gameMode, this.hardcore, difficulty, this.allowCommands, this.gameRules, this.dataPackSettings);
    }
 
