--- a/net/minecraft/entity/effect/StatusEffects.java
+++ b/net/minecraft/entity/effect/StatusEffects.java
@@ -8,6 +8,7 @@
 import net.minecraft.util.registry.Registry;
 import net.minecraft.world.Difficulty;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class StatusEffects {
    public static final StatusEffect SPEED = register(1, "speed", (new StatusEffect(StatusEffectType.BENEFICIAL, 8171462)).addAttributeModifier(EntityAttributes.GENERIC_MOVEMENT_SPEED, "91AEAA56-376B-4498-935B-2F7F68070635", (double)0.2F, EntityAttributeModifier.Operation.MULTIPLY_TOTAL));
    public static final StatusEffect SLOWNESS = register(2, "slowness", (new StatusEffect(StatusEffectType.HARMFUL, 5926017)).addAttributeModifier(EntityAttributes.GENERIC_MOVEMENT_SPEED, "7107DE5E-7CE8-4030-940E-514C1F160890", (double)-0.15F, EntityAttributeModifier.Operation.MULTIPLY_TOTAL));
