--- a/net/minecraft/client/util/ModelIdentifier.java
+++ b/net/minecraft/client/util/ModelIdentifier.java
@@ -1,13 +1,10 @@
 package net.minecraft.client.util;
 
 import java.util.Locale;
-import net.fabricmc.api.EnvType;
-import net.fabricmc.api.Environment;
 import net.minecraft.util.Identifier;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-@Environment(EnvType.CLIENT)
 public class ModelIdentifier extends Identifier {
    private final String variant;
 
