--- a/net/minecraft/entity/FallingBlockEntity.java
+++ b/net/minecraft/entity/FallingBlockEntity.java
@@ -6,7 +6,6 @@
 import net.fabricmc.api.Environment;
 import net.minecraft.block.AnvilBlock;
 import net.minecraft.block.Block;
-import net.minecraft.block.BlockEntityProvider;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.ConcretePowderBlock;
@@ -154,7 +153,7 @@
                               ((FallingBlock)block).onLanding(this.world, blockpos1, this.block, blockstate, this);
                            }
 
-                           if (this.blockEntityData != null && block instanceof BlockEntityProvider) {
+                           if (this.blockEntityData != null && this.block.hasTileEntity()) {
                               BlockEntity tileentity = this.world.getBlockEntity(blockpos1);
                               if (tileentity != null) {
                                  CompoundTag compoundnbt = tileentity.toTag(new CompoundTag());
