--- a/net/minecraft/entity/LivingEntity.java
+++ b/net/minecraft/entity/LivingEntity.java
@@ -115,7 +115,9 @@
 public abstract class LivingEntity extends Entity {
    private static final UUID SPRINTING_SPEED_BOOST_ID = UUID.fromString("662A6B8D-DA3E-4C1C-8813-96EA6097278D");
    private static final UUID SOUL_SPEED_BOOST_ID = UUID.fromString("87f46a96-686f-4796-b035-22e16ee9e038");
+   private static final UUID SLOW_FALLING_ID = UUID.fromString("A5B6CF2A-2F7C-31EF-9022-7C3E7D5E6ABA");
    private static final EntityAttributeModifier SPRINTING_SPEED_BOOST = new EntityAttributeModifier(SPRINTING_SPEED_BOOST_ID, "Sprinting speed boost", (double)0.3F, EntityAttributeModifier.Operation.MULTIPLY_TOTAL);
+   private static final EntityAttributeModifier SLOW_FALLING = new EntityAttributeModifier(SLOW_FALLING_ID, "Slow falling acceleration reduction", -0.07, EntityAttributeModifier.Operation.ADDITION); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
    protected static final TrackedData<Byte> LIVING_FLAGS = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.BYTE);
    private static final TrackedData<Float> HEALTH = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.FLOAT);
    private static final TrackedData<Integer> POTION_SWIRLS_COLOR = DataTracker.registerData(LivingEntity.class, TrackedDataHandlerRegistry.INTEGER);
@@ -243,7 +245,7 @@
    }
 
    public static DefaultAttributeContainer.Builder createLivingAttributes() {
-      return DefaultAttributeContainer.builder().add(EntityAttributes.GENERIC_MAX_HEALTH).add(EntityAttributes.GENERIC_KNOCKBACK_RESISTANCE).add(EntityAttributes.GENERIC_MOVEMENT_SPEED).add(EntityAttributes.GENERIC_ARMOR).add(EntityAttributes.GENERIC_ARMOR_TOUGHNESS);
+      return DefaultAttributeContainer.builder().add(EntityAttributes.GENERIC_MAX_HEALTH).add(EntityAttributes.GENERIC_KNOCKBACK_RESISTANCE).add(EntityAttributes.GENERIC_MOVEMENT_SPEED).add(EntityAttributes.GENERIC_ARMOR).add(EntityAttributes.GENERIC_ARMOR_TOUGHNESS).add(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).add(net.minecraftforge.common.ForgeMod.NAMETAG_DISTANCE.get()).add(net.minecraftforge.common.ForgeMod.ENTITY_GRAVITY.get());
    }
 
    protected void fall(double y, boolean onGroundIn, BlockState state, BlockPos pos) {
@@ -258,9 +260,10 @@
 
       if (!this.world.isClient && this.fallDistance > 3.0F && onGroundIn) {
          float f = (float)MathHelper.ceil(this.fallDistance - 3.0F);
-         if (!state.isAir()) {
+         if (!state.isAir(world, pos)) {
             double d0 = Math.min((double)(0.2F + f / 15.0F), 2.5D);
             int i = (int)(150.0D * d0);
+            if (!state.addLandingEffects((ServerWorld)this.world, pos, state, this, i))
             ((ServerWorld)this.world).spawnParticles(new BlockStateParticleEffect(ParticleTypes.BLOCK, state), this.getX(), this.getY(), this.getZ(), i, 0.0D, 0.0D, 0.0D, (double)0.15F);
          }
       }
@@ -328,7 +331,7 @@
                }
             }
 
-            if (!this.world.isClient && this.hasVehicle() && this.getVehicle() != null && !this.getVehicle().canBeRiddenInWater()) {
+            if (!this.world.isClient && this.hasVehicle() && this.getVehicle() != null && !this.getVehicle().canBeRiddenInWater(this)) {
                this.stopRiding();
             }
          } else if (this.getAir() < this.getMaxAir()) {
@@ -473,7 +476,7 @@
    protected void updatePostDeath() {
       ++this.deathTime;
       if (this.deathTime == 20) {
-         this.remove();
+         this.remove(this instanceof net.minecraft.server.network.ServerPlayerEntity); //Forge keep data until we revive player
 
          for(int i = 0; i < 20; ++i) {
             double d0 = this.random.nextGaussian() * 0.02D;
@@ -666,7 +669,7 @@
             if (!effectinstance.update(this, () -> {
                this.onStatusEffectUpgraded(effectinstance, true);
             })) {
-               if (!this.world.isClient) {
+               if (!this.world.isClient && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionExpiryEvent(this, effectinstance))) {
                   iterator.remove();
                   this.onStatusEffectRemoved(effectinstance);
                }
@@ -715,8 +718,10 @@
          this.setInvisible(false);
       } else {
          Collection<StatusEffectInstance> collection = this.activeStatusEffects.values();
-         this.dataTracker.set(POTION_SWIRLS_AMBIENT, containsOnlyAmbientEffects(collection));
-         this.dataTracker.set(POTION_SWIRLS_COLOR, PotionUtil.getColor(collection));
+         net.minecraftforge.event.entity.living.PotionColorCalculationEvent event = new net.minecraftforge.event.entity.living.PotionColorCalculationEvent(this, PotionUtil.getColor(collection), containsOnlyAmbientEffects(collection), collection);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         this.dataTracker.set(POTION_SWIRLS_AMBIENT, event.areParticlesHidden());
+         this.dataTracker.set(POTION_SWIRLS_COLOR, event.getColor());
          this.setInvisible(this.hasStatusEffect(StatusEffects.INVISIBILITY));
       }
 
@@ -780,7 +785,9 @@
 
          boolean flag;
          for(flag = false; iterator.hasNext(); flag = true) {
-            this.onStatusEffectRemoved(iterator.next());
+            StatusEffectInstance effect = iterator.next();
+            if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effect))) continue;
+            this.onStatusEffectRemoved(effect);
             iterator.remove();
          }
 
@@ -810,6 +817,7 @@
          return false;
       } else {
          StatusEffectInstance effectinstance = this.activeStatusEffects.get(effectInstanceIn.getEffectType());
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionAddedEvent(this, effectinstance, effectInstanceIn));
          if (effectinstance == null) {
             this.activeStatusEffects.put(effectInstanceIn.getEffectType(), effectInstanceIn);
             this.onStatusEffectApplied(effectInstanceIn);
@@ -824,6 +832,9 @@
    }
 
    public boolean canHaveStatusEffect(StatusEffectInstance potioneffectIn) {
+      net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, potioneffectIn);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+      if (event.getResult() != net.minecraftforge.eventbus.api.Event.Result.DEFAULT) return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
       if (this.getGroup() == EntityGroup.UNDEAD) {
          StatusEffect effect = potioneffectIn.getEffectType();
          if (effect == StatusEffects.REGENERATION || effect == StatusEffects.POISON) {
@@ -857,6 +868,7 @@
    }
 
    public boolean removeStatusEffect(StatusEffect effectIn) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effectIn))) return false;
       StatusEffectInstance effectinstance = this.removeStatusEffectInternal(effectIn);
       if (effectinstance != null) {
          this.onStatusEffectRemoved(effectinstance);
@@ -893,6 +905,8 @@
    }
 
    public void heal(float healAmount) {
+      healAmount = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, healAmount);
+      if (healAmount <= 0) return;
       float f = this.getHealth();
       if (f > 0.0F) {
          this.setHealth(f + healAmount);
@@ -913,6 +927,7 @@
    }
 
    public boolean damage(DamageSource source, float amount) {
+      if (!net.minecraftforge.common.ForgeHooks.onLivingAttack(this, source, amount)) return false;
       if (this.isInvulnerableTo(source)) {
          return false;
       } else if (this.world.isClient) {
@@ -979,8 +994,8 @@
             if (entity1 instanceof PlayerEntity) {
                this.playerHitTimer = 100;
                this.attackingPlayer = (PlayerEntity)entity1;
-            } else if (entity1 instanceof WolfEntity) {
-               WolfEntity wolfentity = (WolfEntity)entity1;
+            } else if (entity1 instanceof net.minecraft.entity.passive.TameableEntity) {
+               net.minecraft.entity.passive.TameableEntity wolfentity = (net.minecraft.entity.passive.TameableEntity)entity1;
                if (wolfentity.isTamed()) {
                   this.playerHitTimer = 100;
                   LivingEntity livingentity = wolfentity.getOwner();
@@ -1163,6 +1178,7 @@
    }
 
    public void onDeath(DamageSource cause) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, cause)) return;
       if (!this.removed && !this.dead) {
          Entity entity = cause.getAttacker();
          LivingEntity livingentity = this.getPrimeAdversary();
@@ -1194,10 +1210,10 @@
       if (!this.world.isClient) {
          boolean flag = false;
          if (entitySource instanceof WitherEntity) {
-            if (this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)) {
                BlockPos blockpos = this.getBlockPos();
                BlockState blockstate = Blocks.WITHER_ROSE.getDefaultState();
-               if (this.world.getBlockState(blockpos).isAir() && blockstate.canPlaceAt(this.world, blockpos)) {
+               if (this.world.isAir(blockpos) && blockstate.canPlaceAt(this.world, blockpos)) {
                   this.world.setBlockState(blockpos, blockstate, 3);
                   flag = true;
                }
@@ -1214,13 +1230,10 @@
 
    protected void drop(DamageSource damageSourceIn) {
       Entity entity = damageSourceIn.getAttacker();
-      int i;
-      if (entity instanceof PlayerEntity) {
-         i = EnchantmentHelper.getLooting((LivingEntity)entity);
-      } else {
-         i = 0;
-      }
 
+      int i = net.minecraftforge.common.ForgeHooks.getLootingLevel(this, entity, damageSourceIn);
+      this.captureDrops(new java.util.ArrayList<>());
+
       boolean flag = this.playerHitTimer > 0;
       if (this.shouldDropLoot() && this.world.getGameRules().getBoolean(GameRules.DO_MOB_LOOT)) {
          this.dropLoot(damageSourceIn, flag);
@@ -1229,6 +1242,10 @@
 
       this.dropInventory();
       this.dropXp();
+
+      Collection<ItemEntity> drops = captureDrops(null);
+      if (!net.minecraftforge.common.ForgeHooks.onLivingDrops(this, damageSourceIn, drops, i, playerHitTimer > 0))
+         drops.forEach(e -> world.spawnEntity(e));
    }
 
    protected void dropInventory() {
@@ -1238,6 +1255,7 @@
       if (!this.world.isClient && (this.shouldAlwaysDropXp() || this.playerHitTimer > 0 && this.canDropLootAndXp() && this.world.getGameRules().getBoolean(GameRules.DO_MOB_LOOT))) {
          int i = this.getCurrentExperience(this.attackingPlayer);
 
+         i = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.attackingPlayer, i);
          while(i > 0) {
             int j = ExperienceOrbEntity.roundToOrbSize(i);
             i -= j;
@@ -1258,7 +1276,8 @@
       Identifier resourcelocation = this.getLootTable();
       LootTable loottable = this.world.getServer().getLootManager().getTable(resourcelocation);
       LootContext.Builder lootcontext$builder = this.getLootContextBuilder(attackedRecently, damageSourceIn);
-      loottable.generateLoot(lootcontext$builder.build(LootContextTypes.ENTITY), this::dropStack);
+      LootContext ctx = lootcontext$builder.build(LootContextTypes.ENTITY);
+      loottable.generateLoot(ctx).forEach(this::dropStack);
    }
 
    protected LootContext.Builder getLootContextBuilder(boolean attackedRecently, DamageSource damageSourceIn) {
@@ -1271,6 +1290,11 @@
    }
 
    public void takeKnockback(float strength, double ratioX, double ratioZ) {
+      net.minecraftforge.event.entity.living.LivingKnockBackEvent event = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(this, strength, ratioX, ratioZ);
+      if(event.isCanceled()) return;
+      strength = event.getStrength();
+      ratioX = event.getRatioX();
+      ratioZ = event.getRatioZ();
       strength = (float)((double)strength * (1.0D - this.getAttributeValue(EntityAttributes.GENERIC_KNOCKBACK_RESISTANCE)));
       if (!(strength <= 0.0F)) {
          this.velocityDirty = true;
@@ -1320,16 +1344,7 @@
       } else {
          BlockPos blockpos = this.getBlockPos();
          BlockState blockstate = this.getBlockState();
-         Block block = blockstate.getBlock();
-         if (block.isIn(BlockTags.CLIMBABLE)) {
-            this.climbingPos = Optional.of(blockpos);
-            return true;
-         } else if (block instanceof TrapdoorBlock && this.canEnterTrapdoor(blockpos, blockstate)) {
-            this.climbingPos = Optional.of(blockpos);
-            return true;
-         } else {
-            return false;
-         }
+         return net.minecraftforge.common.ForgeHooks.isLivingOnLadder(blockstate, world, blockpos, this);
       }
    }
 
@@ -1353,6 +1368,11 @@
    }
 
    public boolean handleFallDamage(float distance, float damageMultiplier) {
+      float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, distance, damageMultiplier);
+      if (ret == null) return false;
+      distance = ret[0];
+      damageMultiplier = ret[1];
+
       boolean flag = super.handleFallDamage(distance, damageMultiplier);
       int i = this.computeFallDamage(distance, damageMultiplier);
       if (i > 0) {
@@ -1376,9 +1396,10 @@
          int i = MathHelper.floor(this.getX());
          int j = MathHelper.floor(this.getY() - (double)0.2F);
          int k = MathHelper.floor(this.getZ());
-         BlockState blockstate = this.world.getBlockState(new BlockPos(i, j, k));
-         if (!blockstate.isAir()) {
-            BlockSoundGroup soundtype = blockstate.getSoundGroup();
+         BlockPos pos = new BlockPos(i, j, k);
+         BlockState blockstate = this.world.getBlockState(pos);
+         if (!blockstate.isAir(this.world, pos)) {
+            BlockSoundGroup soundtype = blockstate.getSoundType(world, pos, this);
             this.playSound(soundtype.getFallSound(), soundtype.getVolume() * 0.5F, soundtype.getPitch() * 0.75F);
          }
 
@@ -1446,6 +1467,8 @@
 
    protected void applyDamage(DamageSource damageSrc, float damageAmount) {
       if (!this.isInvulnerableTo(damageSrc)) {
+         damageAmount = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, damageSrc, damageAmount);
+         if (damageAmount <= 0) return;
          damageAmount = this.applyArmorToDamage(damageSrc, damageAmount);
          damageAmount = this.applyEnchantmentsToDamage(damageSrc, damageAmount);
          float f2 = Math.max(damageAmount - this.getAbsorptionAmount(), 0.0F);
@@ -1455,10 +1478,11 @@
             ((ServerPlayerEntity)damageSrc.getAttacker()).increaseStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(f * 10.0F));
          }
 
+         f2 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, damageSrc, f2);
          if (f2 != 0.0F) {
             float f1 = this.getHealth();
-            this.setHealth(f1 - f2);
             this.getDamageTracker().onDamage(damageSrc, f1, f2);
+            this.setHealth(f1 - f2); // Forge: moved to fix MC-121048
             this.setAbsorptionAmount(this.getAbsorptionAmount() - f2);
          }
       }
@@ -1512,6 +1536,8 @@
    }
 
    public void swingHand(Hand handIn, boolean updateSelf) {
+      ItemStack stack = this.getStackInHand(handIn);
+      if (!stack.isEmpty() && stack.onEntitySwing(this)) return;
       if (!this.handSwinging || this.handSwingTicks >= this.getHandSwingDuration() / 2 || this.handSwingTicks < 0) {
          this.handSwingTicks = -1;
          this.handSwinging = true;
@@ -1846,15 +1872,16 @@
       }
 
       this.velocityDirty = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    @Environment(EnvType.CLIENT)
    protected void knockDownwards() {
-      this.setVelocity(this.getVelocity().add(0.0D, (double)-0.04F, 0.0D));
+      this.setVelocity(this.getVelocity().add(0.0D, (double)-0.04F * this.getAttributeInstance(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
    }
 
    protected void swimUpward(net.minecraft.tag.Tag<Fluid> fluidTag) {
-      this.setVelocity(this.getVelocity().add(0.0D, (double)0.04F, 0.0D));
+      this.setVelocity(this.getVelocity().add(0.0D, (double)0.04F * this.getAttributeInstance(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue(), 0.0D));
    }
 
    protected float getBaseMovementSpeedMultiplier() {
@@ -1868,11 +1895,15 @@
    public void travel(Vec3d travelVector) {
       if (this.canMoveVoluntarily() || this.isLogicalSideForUpdatingMovement()) {
          double d0 = 0.08D;
+         EntityAttributeInstance gravity = this.getAttributeInstance(net.minecraftforge.common.ForgeMod.ENTITY_GRAVITY.get());
          boolean flag = this.getVelocity().y <= 0.0D;
          if (flag && this.hasStatusEffect(StatusEffects.SLOW_FALLING)) {
-            d0 = 0.01D;
+            if (!gravity.hasModifier(SLOW_FALLING)) gravity.addTemporaryModifier(SLOW_FALLING);
             this.fallDistance = 0.0F;
+         } else if (gravity.hasModifier(SLOW_FALLING)) {
+            gravity.removeModifier(SLOW_FALLING);
          }
+         d0 = gravity.getValue();
 
          FluidState fluidstate = this.world.getFluidState(this.getBlockPos());
          if (this.isTouchingWater() && this.method_29920() && !this.canWalkOnFluid(fluidstate.getFluid())) {
@@ -1897,6 +1928,7 @@
                f5 = 0.96F;
             }
 
+            f6 *= (float)this.getAttributeInstance(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).getValue();
             this.updateVelocity(f6, travelVector);
             this.move(MovementType.SELF, this.getVelocity());
             Vec3d vector3d6 = this.getVelocity();
@@ -1975,7 +2007,7 @@
             }
          } else {
             BlockPos blockpos = this.getVelocityAffectingPos();
-            float f3 = this.world.getBlockState(blockpos).getBlock().getSlipperiness();
+            float f3 = this.world.getBlockState(this.getVelocityAffectingPos()).getSlipperiness(world, this.getVelocityAffectingPos(), this);
             float f4 = this.onGround ? f3 * 0.91F : 0.91F;
             Vec3d vector3d5 = this.method_26318(travelVector, f3);
             double d2 = vector3d5.y;
@@ -2047,7 +2079,7 @@
          double d0 = MathHelper.clamp(p_213362_1_.x, (double)-0.15F, (double)0.15F);
          double d1 = MathHelper.clamp(p_213362_1_.z, (double)-0.15F, (double)0.15F);
          double d2 = Math.max(p_213362_1_.y, (double)-0.15F);
-         if (d2 < 0.0D && !this.getBlockState().isOf(Blocks.SCAFFOLDING) && this.isHoldingOntoLadder() && this instanceof PlayerEntity) {
+         if (d2 < 0.0D && !this.getBlockState().isScaffolding(this) && this.isHoldingOntoLadder() && this instanceof PlayerEntity) {
             d2 = 0.0D;
          }
 
@@ -2075,6 +2107,7 @@
    }
 
    public void tick() {
+      if (net.minecraftforge.common.ForgeHooks.onLivingUpdate(this)) return;
       super.tick();
       this.tickActiveItemStack();
       this.updateLeaningPitch();
@@ -2230,6 +2263,7 @@
 
          ItemStack itemstack1 = this.getEquippedStack(equipmentslottype);
          if (!ItemStack.areEqual(itemstack1, itemstack)) {
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.LivingEquipmentChangeEvent(this, equipmentslottype, itemstack, itemstack1));
             if (map == null) {
                map = Maps.newEnumMap(EquipmentSlot.class);
             }
@@ -2434,6 +2468,8 @@
       boolean flag = this.getFlag(7);
       if (flag && !this.onGround && !this.hasVehicle() && !this.hasStatusEffect(StatusEffects.LEVITATION)) {
          ItemStack itemstack = this.getEquippedStack(EquipmentSlot.CHEST);
+         flag = itemstack.canElytraFly(this) && itemstack.elytraFlightTick(this, this.roll);
+         if (false) //Forge: Moved to ElytraItem
          if (itemstack.getItem() == Items.ELYTRA && ElytraItem.isUsable(itemstack)) {
             flag = true;
             if (!this.world.isClient && (this.roll + 1) % 20 == 0) {
@@ -2666,8 +2702,16 @@
 
    private void tickActiveItemStack() {
       if (this.isUsingItem()) {
-         if (ItemStack.areItemsEqual(this.getStackInHand(this.getActiveHand()), this.activeItemStack)) {
-            this.activeItemStack = this.getStackInHand(this.getActiveHand());
+         ItemStack itemStack = this.getStackInHand(this.getActiveHand());
+         if (net.minecraftforge.common.ForgeHooks.canContinueUsing(this.activeItemStack, itemStack)) this.activeItemStack = itemStack;
+         if (itemStack == this.activeItemStack) {
+
+            if (!this.activeItemStack.isEmpty()) {
+              itemUseTimeLeft = net.minecraftforge.event.ForgeEventFactory.onItemUseTick(this, activeItemStack, itemUseTimeLeft);
+              if (itemUseTimeLeft > 0)
+                 activeItemStack.onUsingTick(this, itemUseTimeLeft);
+            }
+
             this.activeItemStack.usageTick(this.world, this, this.getItemUseTimeLeft());
             if (this.shouldSpawnConsumptionEffects()) {
                this.spawnConsumptionEffects(this.activeItemStack, 5);
@@ -2715,8 +2759,10 @@
    public void setCurrentHand(Hand hand) {
       ItemStack itemstack = this.getStackInHand(hand);
       if (!itemstack.isEmpty() && !this.isUsingItem()) {
+         int duration = net.minecraftforge.event.ForgeEventFactory.onItemUseStart(this, itemstack, itemstack.getMaxUseTime());
+         if (duration <= 0) return;
          this.activeItemStack = itemstack;
-         this.itemUseTimeLeft = itemstack.getMaxUseTime();
+         this.itemUseTimeLeft = duration;
          if (!this.world.isClient) {
             this.setLivingFlag(1, true);
             this.setLivingFlag(2, hand == Hand.OFF_HAND);
@@ -2776,6 +2822,9 @@
          vector3d1 = vector3d1.rotateX(-this.pitch * ((float)Math.PI / 180F));
          vector3d1 = vector3d1.rotateY(-this.yaw * ((float)Math.PI / 180F));
          vector3d1 = vector3d1.add(this.getX(), this.getEyeY(), this.getZ());
+         if (this.world instanceof ServerWorld) //Forge: Fix MC-2518 spawnParticle is nooped on server, need to use server specific variant
+             ((ServerWorld)this.world).spawnParticles(new ItemStackParticleEffect(ParticleTypes.ITEM, stack), vector3d1.x, vector3d1.y, vector3d1.z, 1, vector3d.x, vector3d.y + 0.05D, vector3d.z, 0.0D);
+         else
          this.world.addParticle(new ItemStackParticleEffect(ParticleTypes.ITEM, stack), vector3d1.x, vector3d1.y, vector3d1.z, vector3d.x, vector3d.y + 0.05D, vector3d.z);
       }
 
@@ -2788,7 +2837,8 @@
       } else {
          if (!this.activeItemStack.isEmpty() && this.isUsingItem()) {
             this.spawnConsumptionEffects(this.activeItemStack, 16);
-            ItemStack itemstack = this.activeItemStack.finishUsing(this.world, this);
+            ItemStack copy = this.activeItemStack.copy();
+            ItemStack itemstack = net.minecraftforge.event.ForgeEventFactory.onItemUseFinish(this, copy, getItemUseTimeLeft(), this.activeItemStack.finishUsing(this.world, this));
             if (itemstack != this.activeItemStack) {
                this.setStackInHand(hand, itemstack);
             }
@@ -2813,7 +2863,11 @@
 
    public void stopUsingItem() {
       if (!this.activeItemStack.isEmpty()) {
+         if (!net.minecraftforge.event.ForgeEventFactory.onUseItemStop(this, activeItemStack, this.getItemUseTimeLeft())) {
+            ItemStack copy = this instanceof PlayerEntity ? activeItemStack.copy() : null;
          this.activeItemStack.onStoppedUsing(this.world, this, this.getItemUseTimeLeft());
+           if (copy != null && activeItemStack.isEmpty()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem((PlayerEntity)this, copy, getActiveHand());
+         }
          if (this.activeItemStack.isUsedOnRelease()) {
             this.tickActiveItemStack();
          }
@@ -2962,8 +3016,8 @@
       }
 
       BlockState blockstate = this.world.getBlockState(pos);
-      if (blockstate.getBlock() instanceof BedBlock) {
-         this.world.setBlockState(pos, blockstate.with(BedBlock.OCCUPIED, Boolean.valueOf(true)), 3);
+      if (blockstate.isBed(world, pos, this)) {
+         blockstate.setBedOccupied(world, pos, this, true);
       }
 
       this.setPose(EntityPose.SLEEPING);
@@ -2979,15 +3033,15 @@
 
    private boolean isSleepingInBed() {
       return this.getSleepingPosition().map((p_241350_1_) -> {
-         return this.world.getBlockState(p_241350_1_).getBlock() instanceof BedBlock;
+         return net.minecraftforge.event.ForgeEventFactory.fireSleepingLocationCheck(this, p_241350_1_);
       }).orElse(false);
    }
 
    public void wakeUp() {
       this.getSleepingPosition().filter(this.world::isChunkLoaded).ifPresent((p_241348_1_) -> {
          BlockState blockstate = this.world.getBlockState(p_241348_1_);
-         if (blockstate.getBlock() instanceof BedBlock) {
-            this.world.setBlockState(p_241348_1_, blockstate.with(BedBlock.OCCUPIED, Boolean.valueOf(false)), 3);
+         if (blockstate.isBed(world, p_241348_1_, this)) {
+            blockstate.setBedOccupied(world, p_241348_1_, this, false);
             Vec3d vector3d1 = BedBlock.findWakeUpPosition(this.getType(), this.world, p_241348_1_, this.yaw).orElseGet(() -> {
                BlockPos blockpos = p_241348_1_.up();
                return new Vec3d((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.1D, (double)blockpos.getZ() + 0.5D);
@@ -3010,7 +3064,9 @@
    @Environment(EnvType.CLIENT)
    public Direction getSleepingDirection() {
       BlockPos blockpos = this.getSleepingPosition().orElse((BlockPos)null);
-      return blockpos != null ? BedBlock.getDirection(this.world, blockpos) : null;
+      if (blockpos == null) return Direction.UP;
+      BlockState state = this.world.getBlockState(blockpos);
+      return !state.isBed(world, blockpos, this) ? Direction.UP : state.getBedDirection(world, blockpos);
    }
 
    public boolean isInsideWall() {
@@ -3080,6 +3136,58 @@
       this.sendEquipmentBreakStatus(p_213334_1_ == Hand.MAIN_HAND ? EquipmentSlot.MAINHAND : EquipmentSlot.OFFHAND);
    }
 
+   /* ==== FORGE START ==== */
+   /***
+    * Removes all potion effects that have curativeItem as a curative item for its effect
+    * @param curativeItem The itemstack we are using to cure potion effects
+    */
+   public boolean curePotionEffects(ItemStack curativeItem) {
+      if (this.world.isClient)
+         return false;
+      boolean ret = false;
+      Iterator<StatusEffectInstance> itr = this.activeStatusEffects.values().iterator();
+      while (itr.hasNext()) {
+         StatusEffectInstance effect = itr.next();
+         if (effect.isCurativeItem(curativeItem) && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.PotionEvent.PotionRemoveEvent(this, effect))) {
+            this.onStatusEffectRemoved(effect);
+            itr.remove();
+            ret = true;
+            this.effectsChanged = true;
+         }
+      }
+      return ret;
+   }
+
+   /**
+    * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+    * currently only used in vanilla code by pigs.
+    *
+    * @param player The player who is riding the entity.
+    * @return If the player should orient the same direction as this entity.
+    */
+   public boolean shouldRiderFaceForward(PlayerEntity player) {
+      return this instanceof net.minecraft.entity.passive.PigEntity;
+   }
+
+   private final net.minecraftforge.common.util.LazyOptional<?>[] handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (this.isAlive() && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) {
+         if (facing == null) return handlers[2].cast();
+         else if (facing.getAxis().isVertical()) return handlers[0].cast();
+         else if (facing.getAxis().isHorizontal()) return handlers[1].cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   protected void invalidateCaps() {
+      super.invalidateCaps();
+      for (int x = 0; x < handlers.length; x++)
+         handlers[x].invalidate();
+   }
+
    @Environment(EnvType.CLIENT)
    public Box getVisibilityBoundingBox() {
       if (this.getEquippedStack(EquipmentSlot.HEAD).getItem() == Items.DRAGON_HEAD) {
