--- a/net/minecraft/entity/SpawnRestriction.java
+++ b/net/minecraft/entity/SpawnRestriction.java
@@ -148,10 +148,26 @@
       boolean test(EntityType<T> p_test_1_, ServerWorldAccess p_test_2_, SpawnReason p_test_3_, BlockPos p_test_4_, Random p_test_5_);
    }
 
-   public static enum Location {
+   public static enum Location implements net.minecraftforge.common.IExtensibleEnum {
       ON_GROUND,
       IN_WATER,
       NO_RESTRICTIONS,
       IN_LAVA;
+
+      public static Location create(String name, net.minecraftforge.common.util.TriPredicate<net.minecraft.world.WorldView, BlockPos, EntityType<? extends MobEntity>> predicate) {
+          throw new IllegalStateException("Enum not extended");
+      }
+
+      private net.minecraftforge.common.util.TriPredicate<net.minecraft.world.WorldView, BlockPos, EntityType<?>> predicate;
+      private Location() { this(null); }
+      private Location(net.minecraftforge.common.util.TriPredicate<net.minecraft.world.WorldView, BlockPos, EntityType<?>> predicate) {
+          this.predicate = predicate;
+      }
+
+      public boolean canSpawnAt(net.minecraft.world.WorldView world, BlockPos pos, EntityType<?> type) {
+          if (this == NO_RESTRICTIONS) return true;
+          if (predicate == null) return net.minecraft.world.SpawnHelper.canSpawnAtBody(this, world, pos, type);
+          return predicate.test(world, pos, type);
+      }
    }
 }
