--- a/net/minecraft/tag/EntityTypeTags.java
+++ b/net/minecraft/tag/EntityTypeTags.java
@@ -16,6 +16,14 @@
       return REQUIRED_TAGS.add(id);
    }
 
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<EntityType<?>> createOptional(Identifier name) {
+       return createOptional(name, null);
+   }
+
+   public static net.minecraftforge.common.Tags.IOptionalNamedTag<EntityType<?>> createOptional(Identifier name, @javax.annotation.Nullable java.util.Set<java.util.function.Supplier<EntityType<?>>> defaults) {
+      return REQUIRED_TAGS.createOptional(name, defaults);
+   }
+
    public static TagGroup<EntityType<?>> getTagGroup() {
       return REQUIRED_TAGS.getGroup();
    }
