--- a/com/mojang/blaze3d/platform/GlStateManager.java
+++ b/com/mojang/blaze3d/platform/GlStateManager.java
@@ -573,10 +573,17 @@
       GL13.glClientActiveTexture(texture);
    }
 
+   /* Stores the last values sent into glMultiTexCoord2f */
+   public static float lastBrightnessX = 0.0f;
+   public static float lastBrightnessY = 0.0f;
    @Deprecated
    public static void multiTexCoords2f(int texture, float s, float t) {
       RenderSystem.assertThread(RenderSystem::isOnRenderThread);
       GL13.glMultiTexCoord2f(texture, s, t);
+      if (texture == GL13.GL_TEXTURE1) {
+          lastBrightnessX = s;
+          lastBrightnessY = t;
+       }
    }
 
    public static void blendFuncSeparateUntracked(int sFactorRGB, int dFactorRGB, int sFactorAlpha, int dFactorAlpha) {
