--- a/net/minecraft/entity/passive/BeeEntity.java
+++ b/net/minecraft/entity/passive/BeeEntity.java
@@ -21,7 +21,6 @@
 import net.minecraft.block.TallPlantBlock;
 import net.minecraft.block.entity.BeehiveBlockEntity;
 import net.minecraft.block.entity.BlockEntity;
-import net.minecraft.block.entity.BlockEntityType;
 import net.minecraft.block.enums.DoubleBlockHalf;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityDimensions;
@@ -427,7 +426,7 @@
          return false;
       } else {
          BlockEntity tileentity = this.world.getBlockEntity(this.hivePos);
-         return tileentity != null && tileentity.getType() == BlockEntityType.BEEHIVE;
+         return tileentity instanceof BeehiveBlockEntity;
       }
    }
 
