--- a/net/minecraft/entity/passive/CatEntity.java
+++ b/net/minecraft/entity/passive/CatEntity.java
@@ -403,7 +403,7 @@
             }
          } else if (this.isBreedingItem(itemstack)) {
             this.eat(p_230254_1_, itemstack);
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_)) {
                this.setOwner(p_230254_1_);
                this.setSitting(true);
                this.world.sendEntityStatus(this, (byte)7);
