--- a/net/minecraft/entity/passive/FoxEntity.java
+++ b/net/minecraft/entity/passive/FoxEntity.java
@@ -758,7 +758,7 @@
       }
 
       protected void eatSweetBerry() {
-         if (FoxEntity.this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+         if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(FoxEntity.this.world, FoxEntity.this)) {
             BlockState blockstate = FoxEntity.this.world.getBlockState(this.targetPos);
             if (blockstate.isOf(Blocks.SWEET_BERRY_BUSH)) {
                int i = blockstate.get(SweetBerryBushBlock.AGE);
@@ -998,6 +998,17 @@
       protected void breed() {
          ServerWorld serverworld = (ServerWorld)this.world;
          FoxEntity foxentity = (FoxEntity)this.animal.createChild(serverworld, this.mate);
+         final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, mate, foxentity);
+         final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         foxentity = (FoxEntity) event.getChild();
+         if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.animal.setBreedingAge(6000);
+            this.mate.setBreedingAge(6000);
+            this.animal.resetLoveTicks();
+            this.mate.resetLoveTicks();
+            return;
+         }
          if (foxentity != null) {
             ServerPlayerEntity serverplayerentity = this.animal.getLovingPlayer();
             ServerPlayerEntity serverplayerentity1 = this.mate.getLovingPlayer();
