--- a/net/minecraft/entity/passive/IronGolemEntity.java
+++ b/net/minecraft/entity/passive/IronGolemEntity.java
@@ -122,9 +122,10 @@
          int i = MathHelper.floor(this.getX());
          int j = MathHelper.floor(this.getY() - (double)0.2F);
          int k = MathHelper.floor(this.getZ());
-         BlockState blockstate = this.world.getBlockState(new BlockPos(i, j, k));
-         if (!blockstate.isAir()) {
-            this.world.addParticle(new BlockStateParticleEffect(ParticleTypes.BLOCK, blockstate), this.getX() + ((double)this.random.nextFloat() - 0.5D) * (double)this.getWidth(), this.getY() + 0.1D, this.getZ() + ((double)this.random.nextFloat() - 0.5D) * (double)this.getWidth(), 4.0D * ((double)this.random.nextFloat() - 0.5D), 0.5D, ((double)this.random.nextFloat() - 0.5D) * 4.0D);
+         BlockPos pos = new BlockPos(i, j, k);
+         BlockState blockstate = this.world.getBlockState(pos);
+         if (!blockstate.isAir(this.world, pos)) {
+            this.world.addParticle(new BlockStateParticleEffect(ParticleTypes.BLOCK, blockstate).setPos(pos), this.getX() + ((double)this.random.nextFloat() - 0.5D) * (double)this.getWidth(), this.getY() + 0.1D, this.getZ() + ((double)this.random.nextFloat() - 0.5D) * (double)this.getWidth(), 4.0D * ((double)this.random.nextFloat() - 0.5D), 0.5D, ((double)this.random.nextFloat() - 0.5D) * 4.0D);
          }
       }
 
