--- a/net/minecraft/entity/passive/OcelotEntity.java
+++ b/net/minecraft/entity/passive/OcelotEntity.java
@@ -164,7 +164,7 @@
       if ((this.temptGoal == null || this.temptGoal.isActive()) && !this.isTrusting() && this.isBreedingItem(itemstack) && p_230254_1_.squaredDistanceTo(this) < 9.0D) {
          this.eat(p_230254_1_, itemstack);
          if (!this.world.isClient) {
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_)) {
                this.setTrusting(true);
                this.showEmoteParticle(true);
                this.world.sendEntityStatus(this, (byte)41);
