--- a/net/minecraft/entity/passive/ParrotEntity.java
+++ b/net/minecraft/entity/passive/ParrotEntity.java
@@ -239,7 +239,7 @@
          }
 
          if (!this.world.isClient) {
-            if (this.random.nextInt(10) == 0) {
+            if (this.random.nextInt(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_)) {
                this.setOwner(p_230254_1_);
                this.world.sendEntityStatus(this, (byte)7);
             } else {
