--- a/net/minecraft/entity/passive/SnowGolemEntity.java
+++ b/net/minecraft/entity/passive/SnowGolemEntity.java
@@ -42,7 +42,7 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class SnowGolemEntity extends GolemEntity implements Shearable, RangedAttackMob {
+public class SnowGolemEntity extends GolemEntity implements Shearable, RangedAttackMob, net.minecraftforge.common.IForgeShearable {
    private static final TrackedData<Byte> SNOW_GOLEM_FLAGS = DataTracker.registerData(SnowGolemEntity.class, TrackedDataHandlerRegistry.BYTE);
 
    public SnowGolemEntity(EntityType<? extends SnowGolemEntity> type, World worldIn) {
@@ -95,7 +95,7 @@
             this.damage(DamageSource.ON_FIRE, 1.0F);
          }
 
-         if (!this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
+         if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)) {
             return;
          }
 
@@ -106,7 +106,7 @@
             j = MathHelper.floor(this.getY());
             k = MathHelper.floor(this.getZ() + (double)((float)(l / 2 % 2 * 2 - 1) * 0.25F));
             BlockPos blockpos = new BlockPos(i, j, k);
-            if (this.world.getBlockState(blockpos).isAir() && this.world.getBiome(blockpos).getTemperature(blockpos) < 0.8F && blockstate.canPlaceAt(this.world, blockpos)) {
+            if (this.world.isAir(blockpos) && this.world.getBiome(blockpos).getTemperature(blockpos) < 0.8F && blockstate.canPlaceAt(this.world, blockpos)) {
                this.world.setBlockState(blockpos, blockstate);
             }
          }
@@ -132,7 +132,7 @@
 
    protected ActionResult interactMob(PlayerEntity p_230254_1_, Hand p_230254_2_) {
       ItemStack itemstack = p_230254_1_.getStackInHand(p_230254_2_);
-      if (itemstack.getItem() == Items.SHEARS && this.isShearable()) {
+      if (false && itemstack.getItem() == Items.SHEARS && this.isShearable()) { //Forge: Moved to onSheared
          this.sheared(SoundCategory.PLAYERS);
          if (!this.world.isClient) {
             itemstack.damage(1, p_230254_1_, (p_213622_1_) -> {
@@ -192,4 +192,20 @@
    public Vec3d method_29919() {
       return new Vec3d(0.0D, (double)(0.75F * this.getStandingEyeHeight()), (double)(this.getWidth() * 0.4F));
    }
+
+   @Override
+   public boolean isShearable(@javax.annotation.Nonnull ItemStack item, World world, BlockPos pos) {
+      return isShearable();
+   }
+
+   @javax.annotation.Nonnull
+   @Override
+   public java.util.List<ItemStack> onSheared(@Nullable PlayerEntity player, @javax.annotation.Nonnull ItemStack item, World world, BlockPos pos, int fortune) {
+      world.playSoundFromEntity(null, this, SoundEvents.ENTITY_SNOW_GOLEM_SHEAR, player == null ? SoundCategory.BLOCKS : SoundCategory.PLAYERS, 1.0F, 1.0F);
+      if (!world.isClient()) {
+         setHasPumpkin(false);
+         return java.util.Collections.singletonList(new ItemStack(Items.CARVED_PUMPKIN));
+      }
+      return java.util.Collections.emptyList();
+   }
 }
