--- a/net/minecraft/entity/passive/WolfEntity.java
+++ b/net/minecraft/entity/passive/WolfEntity.java
@@ -346,7 +346,7 @@
                itemstack.decrement(1);
             }
 
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_)) {
                this.setOwner(p_230254_1_);
                this.navigation.stop();
                this.setTarget((LivingEntity)null);
