--- a/net/minecraft/data/DataGenerator.java
+++ b/net/minecraft/data/DataGenerator.java
@@ -16,10 +16,11 @@
    private final Collection<Path> inputs;
    private final Path output;
    private final List<DataProvider> providers = Lists.newArrayList();
+   private final List<DataProvider> providerView = java.util.Collections.unmodifiableList(providers);
 
    public DataGenerator(Path output, Collection<Path> input) {
       this.output = output;
-      this.inputs = input;
+      this.inputs = Lists.newArrayList(input);
    }
 
    public Collection<Path> getInputs() {
@@ -38,6 +39,7 @@
 
       for(DataProvider idataprovider : this.providers) {
          LOGGER.info("Starting provider: {}", (Object)idataprovider.getName());
+         net.minecraftforge.fml.StartupMessageManager.addModMessage("Generating: " + idataprovider.getName());
          stopwatch1.start();
          idataprovider.run(directorycache);
          stopwatch1.stop();
@@ -53,6 +55,14 @@
       this.providers.add(provider);
    }
 
+   public List<DataProvider> getProviders() {
+       return this.providerView;
+   }
+
+   public void addInput(Path value) {
+      this.inputs.add(value);
+   }
+
    static {
       Bootstrap.initialize();
    }
